{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform Statistical Analysis for AMR genes\n",
    "\n",
    "This analysis is based on: https://www.kaggle.com/code/hamelg/python-for-data-25-chi-squared-tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get low risk patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a function to read data from the FHIR server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "\n",
    "\n",
    "def get(url):\n",
    "    response = requests.get(\n",
    "        url=url,\n",
    "        headers={\"Content-Type\": \"application/fhir+json\", \"authentication\": \"mjRmoNGW6klxaClkKhEkqi7HVYwx6NTH\"},\n",
    "    )\n",
    "    return response\n",
    "\n",
    "\n",
    "def readData(url):\n",
    "    nextUrl = url\n",
    "    data = []\n",
    "    while(nextUrl):\n",
    "        print('Reading URL: ', nextUrl)\n",
    "        response = get(nextUrl)\n",
    "        responseText = json.loads(response.text)\n",
    "        data.append(responseText)\n",
    "        nextUrl = None\n",
    "        if 'link' in responseText:\n",
    "            for link in responseText['link']:\n",
    "                if link['relation'] == 'next':\n",
    "                    nextUrl = link['url']\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain high risk patient ids from FHIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading URL:  http://10.172.235.4:8080/fhir/Patient?_has:RiskAssessment:subject:probability=ge0.0&_has:RiskAssessment:subject:probability=le0.1\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=20&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=40&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=60&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=80&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=100&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=120&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=140&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=160&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=180&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=200&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=220&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=240&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=260&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=280&_count=20&_pretty=true&_bundletype=searchset\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=bec567c3-d73b-464c-97f0-7ec1ed16f4fc&_getpagesoffset=300&_count=20&_pretty=true&_bundletype=searchset\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['2092159',\n",
       " '1710435',\n",
       " '1023735',\n",
       " '745962',\n",
       " '2207219',\n",
       " '2144610',\n",
       " '2301151',\n",
       " '2125063',\n",
       " '1855696',\n",
       " '2190548',\n",
       " '2150228',\n",
       " '2142899',\n",
       " '2263467',\n",
       " '640993',\n",
       " '2198232',\n",
       " '1922539',\n",
       " '2228090',\n",
       " '2185151',\n",
       " '677694',\n",
       " '2271689',\n",
       " '2297160',\n",
       " '2115221',\n",
       " '2199146',\n",
       " '2217919',\n",
       " '2218413',\n",
       " '1584203',\n",
       " '2239222',\n",
       " '2208540',\n",
       " '2233335',\n",
       " '2126898',\n",
       " '2454570',\n",
       " '1012033',\n",
       " '2109854',\n",
       " '2105752',\n",
       " '2080376',\n",
       " '2130121',\n",
       " '360621',\n",
       " '1356357',\n",
       " '2161817',\n",
       " '2085279',\n",
       " '2191641',\n",
       " '2440791',\n",
       " '2133327',\n",
       " '2105813',\n",
       " '2248990',\n",
       " '2384522',\n",
       " '2374110',\n",
       " '2164122',\n",
       " '2526936',\n",
       " '1646600',\n",
       " '2501276',\n",
       " '2141952',\n",
       " '2134257',\n",
       " '1025623',\n",
       " '2421109',\n",
       " '2202499',\n",
       " '2130908',\n",
       " '2221447',\n",
       " '110792',\n",
       " '2144001',\n",
       " '2236567',\n",
       " '2010485',\n",
       " '2252017',\n",
       " '1804919',\n",
       " '1536492',\n",
       " '2164891',\n",
       " '2526402',\n",
       " '2170155',\n",
       " '2213577',\n",
       " '2013664',\n",
       " '2044946',\n",
       " '2205992',\n",
       " '2188910',\n",
       " '2082838',\n",
       " '2183249',\n",
       " '2116081',\n",
       " '2206239',\n",
       " '2110697',\n",
       " '1881022',\n",
       " '2084529',\n",
       " '2223520',\n",
       " '637422',\n",
       " '2042123',\n",
       " '2208200',\n",
       " '2033180',\n",
       " '912828',\n",
       " '2098703',\n",
       " '2146061',\n",
       " '2106126',\n",
       " '2120861',\n",
       " '2126985',\n",
       " '2155539',\n",
       " '2480818',\n",
       " '2116663',\n",
       " '2154045',\n",
       " '2247869',\n",
       " '2092580',\n",
       " '2156000',\n",
       " '2019594',\n",
       " '2122175',\n",
       " '2136431',\n",
       " '2008456',\n",
       " '2087877',\n",
       " '2148889',\n",
       " '2196364',\n",
       " '2439105',\n",
       " '202192',\n",
       " '2186510',\n",
       " '2042164',\n",
       " '41308',\n",
       " '2119435',\n",
       " '2469531',\n",
       " '2107813',\n",
       " '1843548',\n",
       " '890139',\n",
       " '1847691',\n",
       " '2091697',\n",
       " '2205738',\n",
       " '2145022',\n",
       " '2091319',\n",
       " '2037297',\n",
       " '2175656',\n",
       " '2452630',\n",
       " '299125',\n",
       " '1555049',\n",
       " '2126826',\n",
       " '2181057',\n",
       " '2081792',\n",
       " '2045547',\n",
       " '2119654',\n",
       " '2177929',\n",
       " '2100172',\n",
       " '2117184',\n",
       " '2180362',\n",
       " '2020010',\n",
       " '2015999',\n",
       " '2486745',\n",
       " '2202086',\n",
       " '1755211',\n",
       " '2141914',\n",
       " '2200486',\n",
       " '2221815',\n",
       " '2127336',\n",
       " '2109164',\n",
       " '2166951',\n",
       " '328220',\n",
       " '2102099',\n",
       " '2408871',\n",
       " '2253245',\n",
       " '2086417',\n",
       " '2075356',\n",
       " '1673066',\n",
       " '2502910',\n",
       " '2293646',\n",
       " '2200830',\n",
       " '631550',\n",
       " '2216779',\n",
       " '2199560',\n",
       " '2464593',\n",
       " '2130904',\n",
       " '2057069',\n",
       " '1889861',\n",
       " '2270117',\n",
       " '2166007',\n",
       " '681070',\n",
       " '2080648',\n",
       " '2069079',\n",
       " '2198313',\n",
       " '2448944',\n",
       " '1792458',\n",
       " '891293',\n",
       " '2189429',\n",
       " '1011918',\n",
       " '1196189',\n",
       " '635723',\n",
       " '546011',\n",
       " '2231944',\n",
       " '2152424',\n",
       " '2182228',\n",
       " '1056846',\n",
       " '2184520',\n",
       " '2076854',\n",
       " '908851',\n",
       " '2107492',\n",
       " '2249297',\n",
       " '2468015',\n",
       " '1013210',\n",
       " '2131221',\n",
       " '2105392',\n",
       " '2423098',\n",
       " '2103171',\n",
       " '2253645',\n",
       " '296270',\n",
       " '1031912',\n",
       " '2195779',\n",
       " '549608',\n",
       " '1534999',\n",
       " '2083526',\n",
       " '2222200',\n",
       " '2356589',\n",
       " '2123340',\n",
       " '2216836',\n",
       " '2219165',\n",
       " '2336539',\n",
       " '2116995',\n",
       " '2113868',\n",
       " '77432',\n",
       " '2088301',\n",
       " '997341',\n",
       " '2172200',\n",
       " '2101775',\n",
       " '745862',\n",
       " '551009',\n",
       " '2132130',\n",
       " '2031220',\n",
       " '2413187',\n",
       " '634430',\n",
       " '633806',\n",
       " '2148982',\n",
       " '1087409',\n",
       " '2171848',\n",
       " '745573',\n",
       " '352867',\n",
       " '2193278',\n",
       " '2248237',\n",
       " '2161424',\n",
       " '211314',\n",
       " '339352',\n",
       " '2107574',\n",
       " '1017798',\n",
       " '2348296',\n",
       " '2273915',\n",
       " '2160210',\n",
       " '2522101',\n",
       " '2304008',\n",
       " '2023103',\n",
       " '2416190',\n",
       " '2105608',\n",
       " '2223214',\n",
       " '2399366',\n",
       " '2432151',\n",
       " '2122992',\n",
       " '2105605',\n",
       " '2155808',\n",
       " '2153643',\n",
       " '64921',\n",
       " '2085576',\n",
       " '2283682',\n",
       " '821929',\n",
       " '2292417',\n",
       " '2334919',\n",
       " '2091993',\n",
       " '2095728',\n",
       " '2111939',\n",
       " '2149291',\n",
       " '2019734',\n",
       " '2393760',\n",
       " '2124547',\n",
       " '1644679',\n",
       " '2011525',\n",
       " '2097130',\n",
       " '2280800',\n",
       " '1897375',\n",
       " '2257542',\n",
       " '2217661',\n",
       " '2154700',\n",
       " '2181535',\n",
       " '2189763',\n",
       " '2514472',\n",
       " '2469055',\n",
       " '2182776',\n",
       " '2481867',\n",
       " '895279',\n",
       " '2217344',\n",
       " '2186796',\n",
       " '2183779',\n",
       " '684148',\n",
       " '2155815',\n",
       " '2141593',\n",
       " '2161432',\n",
       " '2185888',\n",
       " '1935470',\n",
       " '2123636',\n",
       " '2195751',\n",
       " '297900',\n",
       " '2442667',\n",
       " '2504345',\n",
       " '2140940',\n",
       " '2132697',\n",
       " '1005843',\n",
       " '2171590',\n",
       " '56749',\n",
       " '2199705',\n",
       " '2034458',\n",
       " '320230',\n",
       " '803588',\n",
       " '2108865',\n",
       " '2164770',\n",
       " '2115118',\n",
       " '2220095',\n",
       " '2181175',\n",
       " '1581023',\n",
       " '1014326',\n",
       " '634545',\n",
       " '2143065']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "[lowerRiskScore, higherRiskScore] = [0.0, 0.1]\n",
    "query = 'http://10.172.235.4:8080/fhir/Patient?_has:RiskAssessment:subject:probability=ge' + str(lowerRiskScore) + '&_has:RiskAssessment:subject:probability=le' + str(higherRiskScore)\n",
    "response = readData(query)\n",
    "\n",
    "lowriskPatientIds = list(itertools.chain.from_iterable(list(map(lambda data: list(map(lambda entry: entry['resource']['id'][1:], data['entry'])), response))))\n",
    "lowriskPatientIds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map tube codes for high risk patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tube_code</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>db_ID</th>\n",
       "      <th>pt_age</th>\n",
       "      <th>pt_gender</th>\n",
       "      <th>EPISODE_ID</th>\n",
       "      <th>hospital_admission</th>\n",
       "      <th>hospital_discharge</th>\n",
       "      <th>hospital_in_last_year</th>\n",
       "      <th>genome_species</th>\n",
       "      <th>species_reported</th>\n",
       "      <th>contig_number</th>\n",
       "      <th>length</th>\n",
       "      <th>tube_code_duplicate_or_old</th>\n",
       "      <th>date_of_collection</th>\n",
       "      <th>ID_number</th>\n",
       "      <th>location_additional_02</th>\n",
       "      <th>collected_from_original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH19J072</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>92</td>\n",
       "      <td>5131021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-0997</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH19J074</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus sp. D19</td>\n",
       "      <td>Streptococcus mitis group</td>\n",
       "      <td>84</td>\n",
       "      <td>2022991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-1174</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH19B003</td>\n",
       "      <td>526238.0</td>\n",
       "      <td>567AE</td>\n",
       "      <td>85</td>\n",
       "      <td>Male</td>\n",
       "      <td>640261.0</td>\n",
       "      <td>2019-02-04</td>\n",
       "      <td>2019-02-07</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>93</td>\n",
       "      <td>4942915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-02-03</td>\n",
       "      <td>19-034-1596</td>\n",
       "      <td>A-4WB;GMC</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH20A024</td>\n",
       "      <td>788941.0</td>\n",
       "      <td>3YBDS</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>13194824.0</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>110</td>\n",
       "      <td>5149157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>20-016-2112</td>\n",
       "      <td>S-ED;Emergency</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH20L041</td>\n",
       "      <td>1023735.0</td>\n",
       "      <td>2VJ78</td>\n",
       "      <td>42</td>\n",
       "      <td>Male</td>\n",
       "      <td>14210147.0</td>\n",
       "      <td>2020-12-18</td>\n",
       "      <td>2021-01-05</td>\n",
       "      <td>yes</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>46</td>\n",
       "      <td>2735216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-16</td>\n",
       "      <td>20-351-2856</td>\n",
       "      <td>Emergency Dept;nan</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3112</th>\n",
       "      <td>ALF22L076</td>\n",
       "      <td>639020.0</td>\n",
       "      <td>3S6HL</td>\n",
       "      <td>84</td>\n",
       "      <td>Male</td>\n",
       "      <td>17219899.0</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>2022-12-20</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus vestibularis</td>\n",
       "      <td>Streptococcus salivarius group</td>\n",
       "      <td>107</td>\n",
       "      <td>1938627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>22-349-2150</td>\n",
       "      <td>Ward 4WB</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>ALF22L089</td>\n",
       "      <td>655404.0</td>\n",
       "      <td>6EWBM</td>\n",
       "      <td>67</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteus columbae</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>152</td>\n",
       "      <td>3972219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-17</td>\n",
       "      <td>22-351-1465</td>\n",
       "      <td>Comm/Amb Clinic</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3114</th>\n",
       "      <td>ALF22L138</td>\n",
       "      <td>2572048.0</td>\n",
       "      <td>YM5PJ</td>\n",
       "      <td>61</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>Enterobacter cloacae complex</td>\n",
       "      <td>66</td>\n",
       "      <td>4876179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-27</td>\n",
       "      <td>22-361-0202</td>\n",
       "      <td>Ward 7 East</td>\n",
       "      <td>Hick White Lum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3115</th>\n",
       "      <td>ALF23A102</td>\n",
       "      <td>2054964.0</td>\n",
       "      <td>5DYAH</td>\n",
       "      <td>35</td>\n",
       "      <td>Female</td>\n",
       "      <td>17294287.0</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>2023-02-01</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus constellatus</td>\n",
       "      <td>Streptococcus anginosus group</td>\n",
       "      <td>54</td>\n",
       "      <td>1869828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-11</td>\n",
       "      <td>23-011-3305</td>\n",
       "      <td>ICU</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3116</th>\n",
       "      <td>ALF23B102</td>\n",
       "      <td>2015952.0</td>\n",
       "      <td>57E2W</td>\n",
       "      <td>82</td>\n",
       "      <td>Female</td>\n",
       "      <td>17281239.0</td>\n",
       "      <td>2023-01-08</td>\n",
       "      <td>2023-02-16</td>\n",
       "      <td>no</td>\n",
       "      <td>Acinetobacter colistiniresistens</td>\n",
       "      <td>Acinetobacter species</td>\n",
       "      <td>233</td>\n",
       "      <td>3952992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-15</td>\n",
       "      <td>23-046-3310</td>\n",
       "      <td>ICU</td>\n",
       "      <td>PICC Line White</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3117 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      tube_code  PATIENT_ID  db_ID  pt_age pt_gender  EPISODE_ID  \\\n",
       "0      AH19J072    213972.0  3JNQS      64      Male  12931474.0   \n",
       "1      AH19J074    213972.0  3JNQS      64      Male  12931474.0   \n",
       "2      AH19B003    526238.0  567AE      85      Male    640261.0   \n",
       "3      AH20A024    788941.0  3YBDS      58    Female  13194824.0   \n",
       "4      AH20L041   1023735.0  2VJ78      42      Male  14210147.0   \n",
       "...         ...         ...    ...     ...       ...         ...   \n",
       "3112  ALF22L076    639020.0  3S6HL      84      Male  17219899.0   \n",
       "3113  ALF22L089    655404.0  6EWBM      67    Female         NaN   \n",
       "3114  ALF22L138   2572048.0  YM5PJ      61      Male         NaN   \n",
       "3115  ALF23A102   2054964.0  5DYAH      35    Female  17294287.0   \n",
       "3116  ALF23B102   2015952.0  57E2W      82    Female  17281239.0   \n",
       "\n",
       "     hospital_admission hospital_discharge hospital_in_last_year  \\\n",
       "0            2019-10-17         2019-11-29                   yes   \n",
       "1            2019-10-17         2019-11-29                   yes   \n",
       "2            2019-02-04         2019-02-07                    no   \n",
       "3            2020-01-16         2020-01-19                    no   \n",
       "4            2020-12-18         2021-01-05                   yes   \n",
       "...                 ...                ...                   ...   \n",
       "3112         2022-12-15         2022-12-20                   yes   \n",
       "3113                NaN                NaN                   NaN   \n",
       "3114                NaN                NaN                   NaN   \n",
       "3115         2023-01-12         2023-02-01                   yes   \n",
       "3116         2023-01-08         2023-02-16                    no   \n",
       "\n",
       "                        genome_species                species_reported  \\\n",
       "0                     Escherichia coli                Escherichia coli   \n",
       "1                Streptococcus sp. D19       Streptococcus mitis group   \n",
       "2                     Escherichia coli                Escherichia coli   \n",
       "3                     Escherichia coli                Escherichia coli   \n",
       "4                Staphylococcus aureus           Staphylococcus aureus   \n",
       "...                                ...                             ...   \n",
       "3112        Streptococcus vestibularis  Streptococcus salivarius group   \n",
       "3113                  Proteus columbae                Proteus vulgaris   \n",
       "3114         Enterobacter roggenkampii    Enterobacter cloacae complex   \n",
       "3115        Streptococcus constellatus   Streptococcus anginosus group   \n",
       "3116  Acinetobacter colistiniresistens           Acinetobacter species   \n",
       "\n",
       "      contig_number   length tube_code_duplicate_or_old date_of_collection  \\\n",
       "0                92  5131021                        NaN         2019-10-30   \n",
       "1                84  2022991                        NaN         2019-10-30   \n",
       "2                93  4942915                        NaN         2019-02-03   \n",
       "3               110  5149157                        NaN         2020-01-16   \n",
       "4                46  2735216                        NaN         2020-12-16   \n",
       "...             ...      ...                        ...                ...   \n",
       "3112            107  1938627                        NaN         2022-12-15   \n",
       "3113            152  3972219                        NaN         2022-12-17   \n",
       "3114             66  4876179                        NaN         2022-12-27   \n",
       "3115             54  1869828                        NaN         2023-01-11   \n",
       "3116            233  3952992                        NaN         2023-02-15   \n",
       "\n",
       "        ID_number location_additional_02 collected_from_original  \n",
       "0     19-303-0997             A-7EA;HAEM                   Blood  \n",
       "1     19-303-1174             A-7EA;HAEM        Blood Peripheral  \n",
       "2     19-034-1596              A-4WB;GMC                   Blood  \n",
       "3     20-016-2112         S-ED;Emergency            Blood Venous  \n",
       "4     20-351-2856     Emergency Dept;nan                   Blood  \n",
       "...           ...                    ...                     ...  \n",
       "3112  22-349-2150               Ward 4WB        Blood Peripheral  \n",
       "3113  22-351-1465        Comm/Amb Clinic                   Blood  \n",
       "3114  22-361-0202            Ward 7 East          Hick White Lum  \n",
       "3115  23-011-3305                    ICU            Blood Venous  \n",
       "3116  23-046-3310                    ICU         PICC Line White  \n",
       "\n",
       "[3117 rows x 18 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "mappingDf = pd.read_csv(os.environ['GENOMICS_DATA_BASE'] + '/patient_tube_id_mapping_full.tsv', sep='\\t')\n",
    "mappingDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AH20L041',\n",
       " 'AH20C048',\n",
       " 'AH18K050',\n",
       " 'AH20J071',\n",
       " 'AH19G021',\n",
       " 'AH20A031',\n",
       " 'AH21E085',\n",
       " 'AH21G070',\n",
       " 'AH20I016',\n",
       " 'AH21A081',\n",
       " 'AH21E065',\n",
       " 'AH20I005',\n",
       " 'AH20J009',\n",
       " 'KPN2214',\n",
       " 'AH21B002',\n",
       " 'AH19K005',\n",
       " 'AH21H055',\n",
       " 'AH20G075',\n",
       " 'AH20A054',\n",
       " 'AH21A083',\n",
       " 'AH19J022',\n",
       " 'AH20J052',\n",
       " 'AH21G019',\n",
       " 'AH20I057',\n",
       " 'AH21C022',\n",
       " 'AH20F085',\n",
       " 'AH18J081',\n",
       " 'AH18J080',\n",
       " 'AH18J065',\n",
       " 'AH20F067',\n",
       " 'AH21A011',\n",
       " 'AH18K044',\n",
       " 'AH19L077',\n",
       " 'AH21A043',\n",
       " 'AH20F012',\n",
       " 'AH18K062',\n",
       " 'AH21C014',\n",
       " 'AH21H007',\n",
       " 'AH21H060',\n",
       " 'AH19J028',\n",
       " 'AH20C034',\n",
       " 'AH20H038',\n",
       " 'AH20C060',\n",
       " 'AH19L044',\n",
       " 'AH21E069',\n",
       " 'AH21B039',\n",
       " 'AH21A010',\n",
       " 'AH21E001',\n",
       " 'AH21A075',\n",
       " 'AH20I050',\n",
       " 'AH20A021',\n",
       " 'AH20A022',\n",
       " 'AH21A034',\n",
       " 'AH19I035',\n",
       " 'AH21D030',\n",
       " 'AH19I040',\n",
       " 'AH20L029',\n",
       " 'AH20D059',\n",
       " 'AH20D058',\n",
       " 'AH20J031',\n",
       " 'AH20L010',\n",
       " 'AH21E083',\n",
       " 'AH21H015',\n",
       " 'AH19H053',\n",
       " 'AH20E048',\n",
       " 'AH20H023',\n",
       " 'AH20G045',\n",
       " 'AH20H053',\n",
       " 'AH21G032',\n",
       " 'AH20G069',\n",
       " 'AH20G070',\n",
       " 'AH21B028',\n",
       " 'AH19J034',\n",
       " 'AH20I032',\n",
       " 'AH21D047',\n",
       " 'AH21D049',\n",
       " 'AH20A048',\n",
       " 'AH19L024',\n",
       " 'AH20F086',\n",
       " 'AH21D074',\n",
       " 'AH20D014',\n",
       " 'AH20H013',\n",
       " 'AH20K030',\n",
       " 'AH19L011',\n",
       " 'AH18K008',\n",
       " 'AH21F063',\n",
       " 'AH20H026',\n",
       " 'AH20H027',\n",
       " 'AH21F014',\n",
       " 'AH21F015',\n",
       " 'AH19K013',\n",
       " 'AH20A041',\n",
       " 'AH19J084',\n",
       " 'AH19K006',\n",
       " 'AH20B030',\n",
       " 'AH20B057',\n",
       " 'AH20B058',\n",
       " 'AH20K064',\n",
       " 'AH20L014',\n",
       " 'AH20L012',\n",
       " 'AH19I061',\n",
       " 'AH20E013',\n",
       " 'AH21A048',\n",
       " 'AH20H007',\n",
       " 'AH20A046',\n",
       " 'AH19K071',\n",
       " 'AH20E058',\n",
       " 'AH20E040',\n",
       " 'AH20J034',\n",
       " 'AH21A072',\n",
       " 'AH20H052',\n",
       " 'AH19K018',\n",
       " 'AH19L053',\n",
       " 'AH19L054',\n",
       " 'AH20K061',\n",
       " 'AH19I034',\n",
       " 'AH19F054',\n",
       " 'AH19G028',\n",
       " 'AH19A015',\n",
       " 'AH19A063',\n",
       " 'AH20A004',\n",
       " 'AH20B051',\n",
       " 'AH18K047',\n",
       " 'AH20L044',\n",
       " 'AH19C075',\n",
       " 'AH21H054',\n",
       " 'AH21B038',\n",
       " 'AH19F074',\n",
       " 'AH19F045',\n",
       " 'AH20C026',\n",
       " 'AH19B005',\n",
       " 'AH20L008',\n",
       " 'AH21E027',\n",
       " 'AH19J076',\n",
       " 'AH19J077',\n",
       " 'AH20I027',\n",
       " 'AH21C002',\n",
       " 'AH18K002',\n",
       " 'AH19A042',\n",
       " 'AH19K081',\n",
       " 'AH20K016',\n",
       " 'AH19E069',\n",
       " 'AH21F010',\n",
       " 'AH18L058',\n",
       " 'AH18K016',\n",
       " 'AH20F023',\n",
       " 'AH19C091',\n",
       " 'AH19A038',\n",
       " 'AH20L085',\n",
       " 'AH18K039',\n",
       " 'AH18J053',\n",
       " 'AH19I065',\n",
       " 'KPN1746',\n",
       " 'AH19F031',\n",
       " 'AH19A098',\n",
       " 'AH19D037',\n",
       " 'AH21C041',\n",
       " 'AH19A091',\n",
       " 'AH20H063',\n",
       " 'AH21E079',\n",
       " 'AH20G036',\n",
       " 'AH18J045',\n",
       " 'AH20A069',\n",
       " 'AH19J064',\n",
       " 'AH19B059',\n",
       " 'AH20F039',\n",
       " 'AH20F041',\n",
       " 'AH19I022',\n",
       " 'AH20G066',\n",
       " 'AH21G012',\n",
       " 'AH20E051',\n",
       " 'AH18L025',\n",
       " 'AH20L024',\n",
       " 'AH19B051',\n",
       " 'AH18K040',\n",
       " 'AH19B053',\n",
       " 'AH19A022',\n",
       " 'AH19A068',\n",
       " 'AH19F053',\n",
       " 'AH19D017',\n",
       " 'AH19E059',\n",
       " 'AH20B021',\n",
       " 'AH20I011',\n",
       " 'AH20L035',\n",
       " 'AH21E039',\n",
       " 'AH18K010',\n",
       " 'AH18L018',\n",
       " 'AH19C041',\n",
       " 'AH20A025',\n",
       " 'AH18K073',\n",
       " 'AH20C038',\n",
       " 'AH19L060',\n",
       " 'AH18L057',\n",
       " 'AH20F005',\n",
       " 'AH20F004',\n",
       " 'AH20C005',\n",
       " 'AH18K024',\n",
       " 'AH19E017',\n",
       " 'AH20I045',\n",
       " 'AH20K063',\n",
       " 'AH19D079',\n",
       " 'AH21D015',\n",
       " 'AH19C034',\n",
       " 'AH18J060',\n",
       " 'AH18J063',\n",
       " 'AH19A012',\n",
       " 'AH19G029',\n",
       " 'AH19C025',\n",
       " 'AH19D015',\n",
       " 'AH19L067',\n",
       " 'AH19C012',\n",
       " 'AH20F070',\n",
       " 'AH19B014',\n",
       " 'AH19G060',\n",
       " 'AH18J051',\n",
       " 'AH19C076',\n",
       " 'AH19A007',\n",
       " 'AH19A088',\n",
       " 'AH19H016',\n",
       " 'AH20C041',\n",
       " 'AH19A043',\n",
       " 'AH19B026',\n",
       " 'AH20F033',\n",
       " 'AH20J039',\n",
       " 'AH20L088',\n",
       " 'AH21E078',\n",
       " 'AH21F042',\n",
       " 'AH19G031',\n",
       " 'AH19C070',\n",
       " 'AH19K027',\n",
       " 'AH18L076',\n",
       " 'AH19F068',\n",
       " 'AH20E041',\n",
       " 'AH18K015',\n",
       " 'AH18K065',\n",
       " 'AH18L029',\n",
       " 'AH21F068',\n",
       " 'AH19J035',\n",
       " 'AH20K006',\n",
       " 'AH19K045',\n",
       " 'AH19C087',\n",
       " 'AH18L024',\n",
       " 'AH20I042',\n",
       " 'AH20K011',\n",
       " 'AH19L066',\n",
       " 'AH19H069',\n",
       " 'AH21G025',\n",
       " 'AH21H076',\n",
       " 'AH18K023',\n",
       " 'AH20A072',\n",
       " 'AH19B029',\n",
       " 'AH19D022',\n",
       " 'AH19D004',\n",
       " 'AH19E054',\n",
       " 'AH19I073',\n",
       " 'AH19I072',\n",
       " 'AH20B028',\n",
       " 'AH19G048',\n",
       " 'AH19C020',\n",
       " 'AH20D055',\n",
       " 'AH21B064',\n",
       " 'AH19D055',\n",
       " 'AH19D034',\n",
       " 'AH21C044',\n",
       " 'AH19D043',\n",
       " 'AH19E001',\n",
       " 'AH20L093',\n",
       " 'AH19H078',\n",
       " 'AH21A045',\n",
       " 'AH19A048',\n",
       " 'AH21I005',\n",
       " 'AH21A090',\n",
       " 'AH19C097',\n",
       " 'AH19J068',\n",
       " 'AH18J055',\n",
       " 'AH19I070',\n",
       " 'AH19I071',\n",
       " 'AH19D067',\n",
       " 'AH19F055',\n",
       " 'AH19F072',\n",
       " 'AH21H028',\n",
       " 'AH19H034',\n",
       " 'AH20G044',\n",
       " 'AH19G024',\n",
       " 'AH18J073',\n",
       " 'AH20K029',\n",
       " 'AH18L078',\n",
       " 'AH20K032',\n",
       " 'AH20L079',\n",
       " 'AH19A008',\n",
       " 'AH20K007',\n",
       " 'AH20K040',\n",
       " 'AH19F010',\n",
       " 'AH18J075',\n",
       " 'AH20A038',\n",
       " 'AH19E037',\n",
       " 'AH19I043',\n",
       " 'AH19H023',\n",
       " 'AH19H007',\n",
       " 'AH19H008',\n",
       " 'AH19H009',\n",
       " 'AH19E065',\n",
       " 'AH20I003',\n",
       " 'AH21E034',\n",
       " 'AH20L075',\n",
       " 'AH21C073',\n",
       " 'AH20B069',\n",
       " 'AH19K084',\n",
       " 'AH20I021',\n",
       " 'AH20A064',\n",
       " 'AH19F083',\n",
       " 'AH19J044',\n",
       " 'AH19H062',\n",
       " 'AH19J002',\n",
       " 'AH19K020',\n",
       " 'AH19K030',\n",
       " 'AH19K019',\n",
       " 'AH19L030',\n",
       " 'AH19L032',\n",
       " 'AH20C031',\n",
       " 'AH20I054',\n",
       " 'AH19H024',\n",
       " 'AH19H044',\n",
       " 'AH19H046',\n",
       " 'AH19H051',\n",
       " 'AH20G020',\n",
       " 'AH21E032',\n",
       " 'AH19I050',\n",
       " 'AH19J013',\n",
       " 'AH19K065',\n",
       " 'AH19J036',\n",
       " 'AH19K021',\n",
       " 'AH19L062',\n",
       " 'AH19L027',\n",
       " 'AH19L045',\n",
       " 'AH20C072',\n",
       " 'AH20B052',\n",
       " 'AH20B075',\n",
       " 'AH20B084',\n",
       " 'AH20C042',\n",
       " 'AH20C077',\n",
       " 'AH20J014',\n",
       " 'KPN2241',\n",
       " 'AH20F047',\n",
       " 'AH20G032',\n",
       " 'AH20H004',\n",
       " 'AH20H034',\n",
       " 'AH20I013',\n",
       " 'AH20I039',\n",
       " 'AH20L045',\n",
       " 'AH21D077',\n",
       " 'AH21E019',\n",
       " 'AH20J041',\n",
       " 'AH20K024',\n",
       " 'AH20L073',\n",
       " 'AH20L030',\n",
       " 'AH21A006',\n",
       " 'AH21A007',\n",
       " 'AH21A036',\n",
       " 'AH21A053',\n",
       " 'AH21H043',\n",
       " 'AH21C047',\n",
       " 'AH21C004',\n",
       " 'AH21C012',\n",
       " 'AH21C075',\n",
       " 'AH21D039',\n",
       " 'AH21H048',\n",
       " 'AH21H047',\n",
       " 'AH21F009',\n",
       " 'AH21F012',\n",
       " 'AH21F038',\n",
       " 'AH21G002',\n",
       " 'AH21G026',\n",
       " 'AH21G074',\n",
       " 'AH21H068',\n",
       " 'AH21H029',\n",
       " 'AH21I013',\n",
       " 'AH21J030',\n",
       " 'AH21J061',\n",
       " 'AH21J062',\n",
       " 'AH21K093',\n",
       " 'AH21L003',\n",
       " 'AH22A001',\n",
       " 'ALF22B082',\n",
       " 'ALF22C078',\n",
       " 'ALF22C122',\n",
       " 'ALF22D091',\n",
       " 'ALF22G119',\n",
       " 'ALF22J161',\n",
       " 'ALF22K100',\n",
       " 'ALF22L047']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lowriskTubecodes = list(mappingDf[mappingDf.PATIENT_ID.isin([int(patientId) for patientId in lowriskPatientIds])].tube_code)\n",
    "lowriskTubecodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((8644, 23), (48971, 23))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "lowriskAnnotationsDfList = []\n",
    "controlAnnotationsDfList = []\n",
    "\n",
    "gffDir = os.environ['GENOMICS_DATA_BASE'] + '/amrfinder'\n",
    "\n",
    "for fileName in os.listdir(gffDir):\n",
    "\n",
    "    tubeCode = fileName.split('.')[0].split('_')[0]\n",
    "\n",
    "    amrResultsDf = pd.read_csv(\n",
    "        gffDir + '/' + fileName,\n",
    "        sep='\\t',\n",
    "    )\n",
    "    amrResultsDf['tube_code'] = tubeCode\n",
    "\n",
    "    if tubeCode in lowriskTubecodes:\n",
    "        lowriskAnnotationsDfList.append(amrResultsDf)\n",
    "    else:\n",
    "        controlAnnotationsDfList.append(amrResultsDf)\n",
    "\n",
    "lowriskAnnotationsDf = pd.concat(lowriskAnnotationsDfList, ignore_index=True)\n",
    "controlAnnotationsDf = pd.concat(controlAnnotationsDfList, ignore_index=True)\n",
    "\n",
    "lowriskAnnotationsDf.shape, controlAnnotationsDf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(383, 2597)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(lowriskAnnotationsDf.tube_code.unique()), len(controlAnnotationsDf.tube_code.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform chi-square tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Element type</th>\n",
       "      <th>Element subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VIRULENCE</td>\n",
       "      <td>VIRULENCE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>BIOCIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>METAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>ACID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>HEAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Element type Element subtype\n",
       "0      VIRULENCE       VIRULENCE\n",
       "3            AMR             AMR\n",
       "7         STRESS         BIOCIDE\n",
       "18        STRESS           METAL\n",
       "29        STRESS            ACID\n",
       "220       STRESS            HEAT"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "controlAnnotationsDf[['Element type', 'Element subtype']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annotationType:  VIRULENCE\n",
      "Power_divergenceResult(statistic=23716.75192090004, pvalue=0.0)\n",
      "annotationType:  AMR\n",
      "Power_divergenceResult(statistic=16345.244179286823, pvalue=0.0)\n",
      "annotationType:  STRESS\n",
      "Power_divergenceResult(statistic=9746.498328426738, pvalue=0.0)\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "\n",
    "for annotationType in ['VIRULENCE', 'AMR', 'STRESS']:\n",
    "\n",
    "    print('annotationType: ', annotationType)\n",
    "\n",
    "    highriskGenecountsDf = lowriskAnnotationsDf[lowriskAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'low_risk_genes_count'})\n",
    "\n",
    "    controlGenecountsDf = controlAnnotationsDf[controlAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'control_genes_count'})\n",
    "\n",
    "    mergedGenecountsDf = controlGenecountsDf.merge(\n",
    "        highriskGenecountsDf,\n",
    "        how='left',\n",
    "        on=['Gene symbol']\n",
    "    ).fillna(0)\n",
    "\n",
    "    mergedGenecountsDf['control_genes_proportion'] = mergedGenecountsDf.control_genes_count/mergedGenecountsDf.control_genes_count.sum()\n",
    "\n",
    "    mergedGenecountsDf['expected_genes_count'] = mergedGenecountsDf.control_genes_proportion * mergedGenecountsDf.low_risk_genes_count.sum()\n",
    "\n",
    "    print(stats.chisquare(f_obs= mergedGenecountsDf.low_risk_genes_count, f_exp= mergedGenecountsDf.expected_genes_count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>low_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>arsA</td>\n",
       "      <td>162</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.011670</td>\n",
       "      <td>13.536954</td>\n",
       "      <td>63.064070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>arsB</td>\n",
       "      <td>223</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.016064</td>\n",
       "      <td>18.634203</td>\n",
       "      <td>-259.553889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>arsC</td>\n",
       "      <td>1162</td>\n",
       "      <td>77.0</td>\n",
       "      <td>0.083706</td>\n",
       "      <td>97.098401</td>\n",
       "      <td>20.699003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>arsD</td>\n",
       "      <td>236</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.017000</td>\n",
       "      <td>19.720501</td>\n",
       "      <td>59.433080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>arsD2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.083561</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>terZ</td>\n",
       "      <td>99</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.007132</td>\n",
       "      <td>8.272583</td>\n",
       "      <td>87.911877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>trxLHR</td>\n",
       "      <td>90</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.006483</td>\n",
       "      <td>7.520530</td>\n",
       "      <td>-32.969349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>yfdX1</td>\n",
       "      <td>87</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.006267</td>\n",
       "      <td>7.269846</td>\n",
       "      <td>-10.043599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>yfdX2</td>\n",
       "      <td>89</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.006411</td>\n",
       "      <td>7.436969</td>\n",
       "      <td>-21.017048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>ymgB</td>\n",
       "      <td>838</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.060366</td>\n",
       "      <td>70.024492</td>\n",
       "      <td>85.719282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene symbol  control_genes_count  low_risk_genes_count  \\\n",
       "0         arsA                  162                   5.0   \n",
       "1         arsB                  223                  67.0   \n",
       "2         arsC                 1162                  77.0   \n",
       "3         arsD                  236                   8.0   \n",
       "4        arsD2                    1                   0.0   \n",
       "..         ...                  ...                   ...   \n",
       "82        terZ                   99                   1.0   \n",
       "83      trxLHR                   90                  10.0   \n",
       "84       yfdX1                   87                   8.0   \n",
       "85       yfdX2                   89                   9.0   \n",
       "86        ymgB                  838                  10.0   \n",
       "\n",
       "    control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "0                   0.011670             13.536954              63.064070  \n",
       "1                   0.016064             18.634203            -259.553889  \n",
       "2                   0.083706             97.098401              20.699003  \n",
       "3                   0.017000             19.720501              59.433080  \n",
       "4                   0.000072              0.083561             100.000000  \n",
       "..                       ...                   ...                    ...  \n",
       "82                  0.007132              8.272583              87.911877  \n",
       "83                  0.006483              7.520530             -32.969349  \n",
       "84                  0.006267              7.269846             -10.043599  \n",
       "85                  0.006411              7.436969             -21.017048  \n",
       "86                  0.060366             70.024492              85.719282  \n",
       "\n",
       "[87 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf['percentage_difference'] = (mergedGenecountsDf.expected_genes_count - mergedGenecountsDf.low_risk_genes_count)/mergedGenecountsDf.expected_genes_count * 100\n",
    "mergedGenecountsDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>low_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>mco</td>\n",
       "      <td>21</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.001513</td>\n",
       "      <td>1.754790</td>\n",
       "      <td>-2122.487685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>qacA</td>\n",
       "      <td>10</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.000720</td>\n",
       "      <td>0.835614</td>\n",
       "      <td>-1814.758621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>lmrS</td>\n",
       "      <td>212</td>\n",
       "      <td>304.0</td>\n",
       "      <td>0.015272</td>\n",
       "      <td>17.715027</td>\n",
       "      <td>-1616.057254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cadD</td>\n",
       "      <td>145</td>\n",
       "      <td>205.0</td>\n",
       "      <td>0.010445</td>\n",
       "      <td>12.116410</td>\n",
       "      <td>-1591.920333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>qacR</td>\n",
       "      <td>13</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.000936</td>\n",
       "      <td>1.086299</td>\n",
       "      <td>-1464.946950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>qacJ</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.083561</td>\n",
       "      <td>-1096.724138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>qacC</td>\n",
       "      <td>27</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.001945</td>\n",
       "      <td>2.256159</td>\n",
       "      <td>-653.492976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>merB</td>\n",
       "      <td>8</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000576</td>\n",
       "      <td>0.668492</td>\n",
       "      <td>-348.771552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cadC</td>\n",
       "      <td>6</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.501369</td>\n",
       "      <td>-298.908046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>qacB</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.250684</td>\n",
       "      <td>-298.908046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>arsB</td>\n",
       "      <td>223</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.016064</td>\n",
       "      <td>18.634203</td>\n",
       "      <td>-259.553889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>nreB</td>\n",
       "      <td>23</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.001657</td>\n",
       "      <td>1.921913</td>\n",
       "      <td>-108.125937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>arsR</td>\n",
       "      <td>379</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.027302</td>\n",
       "      <td>31.669788</td>\n",
       "      <td>-105.242926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>trxLHR</td>\n",
       "      <td>90</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.006483</td>\n",
       "      <td>7.520530</td>\n",
       "      <td>-32.969349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>yfdX2</td>\n",
       "      <td>89</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.006411</td>\n",
       "      <td>7.436969</td>\n",
       "      <td>-21.017048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>crcB</td>\n",
       "      <td>30</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.002161</td>\n",
       "      <td>2.506843</td>\n",
       "      <td>-19.672414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>hdeD-GI</td>\n",
       "      <td>85</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.006123</td>\n",
       "      <td>7.102723</td>\n",
       "      <td>-12.632860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>yfdX1</td>\n",
       "      <td>87</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.006267</td>\n",
       "      <td>7.269846</td>\n",
       "      <td>-10.043599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>shsP</td>\n",
       "      <td>91</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>7.604092</td>\n",
       "      <td>-5.206518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>psi-GI</td>\n",
       "      <td>82</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.005907</td>\n",
       "      <td>6.852039</td>\n",
       "      <td>-2.159378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene symbol  control_genes_count  low_risk_genes_count  \\\n",
       "26         mco                   21                  39.0   \n",
       "52        qacA                   10                  16.0   \n",
       "25        lmrS                  212                 304.0   \n",
       "12        cadD                  145                 205.0   \n",
       "62        qacR                   13                  17.0   \n",
       "60        qacJ                    1                   1.0   \n",
       "54        qacC                   27                  17.0   \n",
       "28        merB                    8                   3.0   \n",
       "11        cadC                    6                   2.0   \n",
       "53        qacB                    3                   1.0   \n",
       "1         arsB                  223                  67.0   \n",
       "43        nreB                   23                   4.0   \n",
       "6         arsR                  379                  65.0   \n",
       "83      trxLHR                   90                  10.0   \n",
       "85       yfdX2                   89                   9.0   \n",
       "18        crcB                   30                   3.0   \n",
       "22     hdeD-GI                   85                   8.0   \n",
       "84       yfdX1                   87                   8.0   \n",
       "64        shsP                   91                   8.0   \n",
       "51      psi-GI                   82                   7.0   \n",
       "\n",
       "    control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "26                  0.001513              1.754790           -2122.487685  \n",
       "52                  0.000720              0.835614           -1814.758621  \n",
       "25                  0.015272             17.715027           -1616.057254  \n",
       "12                  0.010445             12.116410           -1591.920333  \n",
       "62                  0.000936              1.086299           -1464.946950  \n",
       "60                  0.000072              0.083561           -1096.724138  \n",
       "54                  0.001945              2.256159            -653.492976  \n",
       "28                  0.000576              0.668492            -348.771552  \n",
       "11                  0.000432              0.501369            -298.908046  \n",
       "53                  0.000216              0.250684            -298.908046  \n",
       "1                   0.016064             18.634203            -259.553889  \n",
       "43                  0.001657              1.921913            -108.125937  \n",
       "6                   0.027302             31.669788            -105.242926  \n",
       "83                  0.006483              7.520530             -32.969349  \n",
       "85                  0.006411              7.436969             -21.017048  \n",
       "18                  0.002161              2.506843             -19.672414  \n",
       "22                  0.006123              7.102723             -12.632860  \n",
       "84                  0.006267              7.269846             -10.043599  \n",
       "64                  0.006555              7.604092              -5.206518  \n",
       "51                  0.005907              6.852039              -2.159378  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf.sort_values(by=['percentage_difference'])[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>low_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>terZ</td>\n",
       "      <td>99</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.007132</td>\n",
       "      <td>8.272583</td>\n",
       "      <td>87.911877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>asr</td>\n",
       "      <td>864</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.062239</td>\n",
       "      <td>72.197090</td>\n",
       "      <td>86.149026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>ymgB</td>\n",
       "      <td>838</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.060366</td>\n",
       "      <td>70.024492</td>\n",
       "      <td>85.719282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>terD</td>\n",
       "      <td>158</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.011382</td>\n",
       "      <td>13.202709</td>\n",
       "      <td>84.851593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>clpK</td>\n",
       "      <td>66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004754</td>\n",
       "      <td>5.515055</td>\n",
       "      <td>81.867816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>emrE</td>\n",
       "      <td>842</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.060654</td>\n",
       "      <td>70.358738</td>\n",
       "      <td>81.523262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>fieF</td>\n",
       "      <td>1422</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.102435</td>\n",
       "      <td>118.824377</td>\n",
       "      <td>75.594233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>pcoE</td>\n",
       "      <td>143</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.010301</td>\n",
       "      <td>11.949287</td>\n",
       "      <td>74.893899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ssmE</td>\n",
       "      <td>39</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002809</td>\n",
       "      <td>3.258896</td>\n",
       "      <td>69.314766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>sdeA</td>\n",
       "      <td>39</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002809</td>\n",
       "      <td>3.258896</td>\n",
       "      <td>69.314766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>smdA</td>\n",
       "      <td>39</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002809</td>\n",
       "      <td>3.258896</td>\n",
       "      <td>69.314766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>smdB</td>\n",
       "      <td>39</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002809</td>\n",
       "      <td>3.258896</td>\n",
       "      <td>69.314766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>silS</td>\n",
       "      <td>274</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.019738</td>\n",
       "      <td>22.895836</td>\n",
       "      <td>65.059149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>silC</td>\n",
       "      <td>272</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.019594</td>\n",
       "      <td>22.728713</td>\n",
       "      <td>64.802231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>silB</td>\n",
       "      <td>272</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.019594</td>\n",
       "      <td>22.728713</td>\n",
       "      <td>64.802231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>qacEdelta1</td>\n",
       "      <td>237</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.017072</td>\n",
       "      <td>19.804063</td>\n",
       "      <td>64.653717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>silF</td>\n",
       "      <td>269</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.019378</td>\n",
       "      <td>22.478029</td>\n",
       "      <td>64.409691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>silP</td>\n",
       "      <td>267</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.019234</td>\n",
       "      <td>22.310906</td>\n",
       "      <td>64.143097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>copB</td>\n",
       "      <td>233</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.016784</td>\n",
       "      <td>19.469817</td>\n",
       "      <td>64.046914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>pcoR</td>\n",
       "      <td>264</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.019017</td>\n",
       "      <td>22.060222</td>\n",
       "      <td>63.735632</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene symbol  control_genes_count  low_risk_genes_count  \\\n",
       "82        terZ                   99                   1.0   \n",
       "7          asr                  864                  10.0   \n",
       "86        ymgB                  838                  10.0   \n",
       "79        terD                  158                   2.0   \n",
       "14        clpK                   66                   1.0   \n",
       "20        emrE                  842                  13.0   \n",
       "21        fieF                 1422                  29.0   \n",
       "48        pcoE                  143                   3.0   \n",
       "75        ssmE                   39                   1.0   \n",
       "63        sdeA                   39                   1.0   \n",
       "73        smdA                   39                   1.0   \n",
       "74        smdB                   39                   1.0   \n",
       "72        silS                  274                   8.0   \n",
       "67        silC                  272                   8.0   \n",
       "66        silB                  272                   8.0   \n",
       "56  qacEdelta1                  237                   7.0   \n",
       "69        silF                  269                   8.0   \n",
       "70        silP                  267                   8.0   \n",
       "16        copB                  233                   7.0   \n",
       "49        pcoR                  264                   8.0   \n",
       "\n",
       "    control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "82                  0.007132              8.272583              87.911877  \n",
       "7                   0.062239             72.197090              86.149026  \n",
       "86                  0.060366             70.024492              85.719282  \n",
       "79                  0.011382             13.202709              84.851593  \n",
       "14                  0.004754              5.515055              81.867816  \n",
       "20                  0.060654             70.358738              81.523262  \n",
       "21                  0.102435            118.824377              75.594233  \n",
       "48                  0.010301             11.949287              74.893899  \n",
       "75                  0.002809              3.258896              69.314766  \n",
       "63                  0.002809              3.258896              69.314766  \n",
       "73                  0.002809              3.258896              69.314766  \n",
       "74                  0.002809              3.258896              69.314766  \n",
       "72                  0.019738             22.895836              65.059149  \n",
       "67                  0.019594             22.728713              64.802231  \n",
       "66                  0.019594             22.728713              64.802231  \n",
       "56                  0.017072             19.804063              64.653717  \n",
       "69                  0.019378             22.478029              64.409691  \n",
       "70                  0.019234             22.310906              64.143097  \n",
       "16                  0.016784             19.469817              64.046914  \n",
       "49                  0.019017             22.060222              63.735632  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf[mergedGenecountsDf.low_risk_genes_count > 0].sort_values(by=['percentage_difference'], ascending=False)[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>low_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>arsD2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.083561</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>arsH</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.083561</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bcrB</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.167123</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bcrC</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.167123</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cadA</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>0.501369</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.083561</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>dpsA</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.167123</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>nirA</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.167123</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>nirB</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.250684</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>nirD</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.250684</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>qacE</td>\n",
       "      <td>41</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>3.426019</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>qacF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.083561</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>qacG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.417807</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>qacG2</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001441</td>\n",
       "      <td>1.671229</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>qacL</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000720</td>\n",
       "      <td>0.835614</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>tcrB</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.417807</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>terB</td>\n",
       "      <td>53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003818</td>\n",
       "      <td>4.428757</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>terC</td>\n",
       "      <td>55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003962</td>\n",
       "      <td>4.595880</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>terE</td>\n",
       "      <td>52</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003746</td>\n",
       "      <td>4.345195</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>terW</td>\n",
       "      <td>44</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003170</td>\n",
       "      <td>3.676704</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene symbol  control_genes_count  low_risk_genes_count  \\\n",
       "4        arsD2                    1                   0.0   \n",
       "5         arsH                    1                   0.0   \n",
       "8         bcrB                    2                   0.0   \n",
       "9         bcrC                    2                   0.0   \n",
       "10        cadA                    6                   0.0   \n",
       "13        chrA                    1                   0.0   \n",
       "19        dpsA                    2                   0.0   \n",
       "40        nirA                    2                   0.0   \n",
       "41        nirB                    3                   0.0   \n",
       "42        nirD                    3                   0.0   \n",
       "55        qacE                   41                   0.0   \n",
       "57        qacF                    1                   0.0   \n",
       "58        qacG                    5                   0.0   \n",
       "59       qacG2                   20                   0.0   \n",
       "61        qacL                   10                   0.0   \n",
       "76        tcrB                    5                   0.0   \n",
       "77        terB                   53                   0.0   \n",
       "78        terC                   55                   0.0   \n",
       "80        terE                   52                   0.0   \n",
       "81        terW                   44                   0.0   \n",
       "\n",
       "    control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "4                   0.000072              0.083561                  100.0  \n",
       "5                   0.000072              0.083561                  100.0  \n",
       "8                   0.000144              0.167123                  100.0  \n",
       "9                   0.000144              0.167123                  100.0  \n",
       "10                  0.000432              0.501369                  100.0  \n",
       "13                  0.000072              0.083561                  100.0  \n",
       "19                  0.000144              0.167123                  100.0  \n",
       "40                  0.000144              0.167123                  100.0  \n",
       "41                  0.000216              0.250684                  100.0  \n",
       "42                  0.000216              0.250684                  100.0  \n",
       "55                  0.002953              3.426019                  100.0  \n",
       "57                  0.000072              0.083561                  100.0  \n",
       "58                  0.000360              0.417807                  100.0  \n",
       "59                  0.001441              1.671229                  100.0  \n",
       "61                  0.000720              0.835614                  100.0  \n",
       "76                  0.000360              0.417807                  100.0  \n",
       "77                  0.003818              4.428757                  100.0  \n",
       "78                  0.003962              4.595880                  100.0  \n",
       "80                  0.003746              4.345195                  100.0  \n",
       "81                  0.003170              3.676704                  100.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf[mergedGenecountsDf.low_risk_genes_count == 0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
