{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform Statistical Analysis for AMR genes\n",
    "\n",
    "This analysis is based on: https://www.kaggle.com/code/hamelg/python-for-data-25-chi-squared-tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get high risk patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a function to read data from the FHIR server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "\n",
    "\n",
    "def get(url):\n",
    "    response = requests.get(\n",
    "        url=url,\n",
    "        headers={\"Content-Type\": \"application/fhir+json\", \"authentication\": \"mjRmoNGW6klxaClkKhEkqi7HVYwx6NTH\"},\n",
    "    )\n",
    "    return response\n",
    "\n",
    "\n",
    "def readData(url):\n",
    "    nextUrl = url\n",
    "    data = []\n",
    "    while(nextUrl):\n",
    "        print('Reading URL: ', nextUrl)\n",
    "        response = get(nextUrl)\n",
    "        responseText = json.loads(response.text)\n",
    "        data.append(responseText)\n",
    "        nextUrl = None\n",
    "        if 'link' in responseText:\n",
    "            for link in responseText['link']:\n",
    "                if link['relation'] == 'next':\n",
    "                    nextUrl = link['url']\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain high risk patient ids from FHIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading URL:  http://10.172.235.4:8080/fhir/Patient?_has:RiskAssessment:subject:probability=ge0.5&_has:RiskAssessment:subject:probability=le1.0\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=0f45a4f1-b5f4-4107-9321-e4d486645d92&_getpagesoffset=20&_count=20&_pretty=true&_bundletype=searchset\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['2202086',\n",
       " '2172200',\n",
       " '546011',\n",
       " '1013210',\n",
       " '2103171',\n",
       " '2013664',\n",
       " '2120861',\n",
       " '2142899',\n",
       " '677694',\n",
       " '2448944',\n",
       " '2150228',\n",
       " '745962',\n",
       " '2153643',\n",
       " '1581023',\n",
       " '2140940',\n",
       " '631550',\n",
       " '2198313',\n",
       " '2161817',\n",
       " '2107492',\n",
       " '2141593',\n",
       " '549608',\n",
       " '2160210',\n",
       " '637422',\n",
       " '2092580',\n",
       " '2156000',\n",
       " '2199705']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "[lowerRiskScore, higherRiskScore] = [0.5, 1.0]\n",
    "query = 'http://10.172.235.4:8080/fhir/Patient?_has:RiskAssessment:subject:probability=ge' + str(lowerRiskScore) + '&_has:RiskAssessment:subject:probability=le' + str(higherRiskScore)\n",
    "response = readData(query)\n",
    "\n",
    "highriskPatientIds = list(itertools.chain.from_iterable(list(map(lambda data: list(map(lambda entry: entry['resource']['id'][1:], data['entry'])), response))))\n",
    "highriskPatientIds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(highriskPatientIds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map tube codes for high risk patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>specimen_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>remap_file</th>\n",
       "      <th>amr_file</th>\n",
       "      <th>fasta_info_file</th>\n",
       "      <th>remap_info_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH18J001</td>\n",
       "      <td>2145906</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH18J002</td>\n",
       "      <td>2154958</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH18J014</td>\n",
       "      <td>652485</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH18J021</td>\n",
       "      <td>2054932</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH18J022</td>\n",
       "      <td>2131734</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>AH21H055</td>\n",
       "      <td>297900</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>AH21H060</td>\n",
       "      <td>745862</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>AH21H068</td>\n",
       "      <td>2526402</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>AH21H076</td>\n",
       "      <td>2180362</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>AH21I005</td>\n",
       "      <td>2195751</td>\n",
       "      <td>/home/vmadmin/workspace/genome_data/remap_file...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "      <td>https://raw.githubusercontent.com/ryashpal/ehr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    specimen_id  patient_id  \\\n",
       "0      AH18J001     2145906   \n",
       "1      AH18J002     2154958   \n",
       "2      AH18J014      652485   \n",
       "3      AH18J021     2054932   \n",
       "4      AH18J022     2131734   \n",
       "..          ...         ...   \n",
       "362    AH21H055      297900   \n",
       "363    AH21H060      745862   \n",
       "364    AH21H068     2526402   \n",
       "365    AH21H076     2180362   \n",
       "366    AH21I005     2195751   \n",
       "\n",
       "                                            remap_file  \\\n",
       "0    /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "1    /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "2    /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "3    /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "4    /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "..                                                 ...   \n",
       "362  /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "363  /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "364  /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "365  /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "366  /home/vmadmin/workspace/genome_data/remap_file...   \n",
       "\n",
       "                                              amr_file  \\\n",
       "0    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "1    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "2    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "3    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "4    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "..                                                 ...   \n",
       "362  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "363  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "364  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "365  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "366  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "\n",
       "                                       fasta_info_file  \\\n",
       "0    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "1    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "2    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "3    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "4    https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "..                                                 ...   \n",
       "362  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "363  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "364  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "365  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "366  https://raw.githubusercontent.com/ryashpal/ehr...   \n",
       "\n",
       "                                       remap_info_file  \n",
       "0    https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "1    https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "2    https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "3    https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "4    https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "..                                                 ...  \n",
       "362  https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "363  https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "364  https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "365  https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "366  https://raw.githubusercontent.com/ryashpal/ehr...  \n",
       "\n",
       "[367 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "mappingDf = pd.read_csv(os.environ['GENOMICS_DATA_BASE'] + '/index_saur.csv')\n",
    "mappingDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AH18J055',\n",
       " 'AH18J081',\n",
       " 'AH18K024',\n",
       " 'AH19A043',\n",
       " 'AH19B014',\n",
       " 'AH19B026',\n",
       " 'AH19C012',\n",
       " 'AH19C070',\n",
       " 'AH19E069',\n",
       " 'AH19F055',\n",
       " 'AH19J028',\n",
       " 'AH19K013',\n",
       " 'AH19L077',\n",
       " 'AH20C005',\n",
       " 'AH20E051',\n",
       " 'AH20F012',\n",
       " 'AH20G044',\n",
       " 'AH20H023',\n",
       " 'AH20I042',\n",
       " 'AH20I057',\n",
       " 'AH20K063',\n",
       " 'AH20L008',\n",
       " 'AH20L085',\n",
       " 'AH21A034',\n",
       " 'AH21A036',\n",
       " 'AH21C022']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highriskTubecodes = list(mappingDf[mappingDf.patient_id.isin([int(patientId) for patientId in highriskPatientIds])].specimen_id)\n",
    "highriskTubecodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((595, 23), (57020, 23))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "highriskAnnotationsDfList = []\n",
    "controlAnnotationsDfList = []\n",
    "\n",
    "gffDir = os.environ['GENOMICS_DATA_BASE'] + '/amrfinder'\n",
    "\n",
    "for fileName in os.listdir(gffDir):\n",
    "\n",
    "    tubeCode = fileName.split('.')[0].split('_')[0]\n",
    "\n",
    "    amrResultsDf = pd.read_csv(\n",
    "        gffDir + '/' + fileName,\n",
    "        sep='\\t',\n",
    "    )\n",
    "    amrResultsDf['tube_code'] = tubeCode\n",
    "\n",
    "    if tubeCode in highriskTubecodes:\n",
    "        highriskAnnotationsDfList.append(amrResultsDf)\n",
    "    else:\n",
    "        controlAnnotationsDfList.append(amrResultsDf)\n",
    "\n",
    "highriskAnnotationsDf = pd.concat(highriskAnnotationsDfList, ignore_index=True)\n",
    "controlAnnotationsDf = pd.concat(controlAnnotationsDfList, ignore_index=True)\n",
    "\n",
    "highriskAnnotationsDf.shape, controlAnnotationsDf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 2954)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(highriskAnnotationsDf.tube_code.unique()), len(controlAnnotationsDf.tube_code.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform chi-square tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Element type</th>\n",
       "      <th>Element subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VIRULENCE</td>\n",
       "      <td>VIRULENCE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>BIOCIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>METAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>ACID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>HEAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Element type Element subtype\n",
       "0      VIRULENCE       VIRULENCE\n",
       "3            AMR             AMR\n",
       "7         STRESS         BIOCIDE\n",
       "18        STRESS           METAL\n",
       "29        STRESS            ACID\n",
       "220       STRESS            HEAT"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "controlAnnotationsDf[['Element type', 'Element subtype']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annotationType:  VIRULENCE\n",
      "55.0 0.05745735167659182 False\n",
      "annotationType:  AMR\n",
      "Not sufficient data for the test\n",
      "annotationType:  STRESS\n",
      "Not sufficient data for the test\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "\n",
    "dfDict = {}\n",
    "\n",
    "for annotationType in ['VIRULENCE', 'AMR', 'STRESS']:\n",
    "# for annotationType in ['AMR']:\n",
    "\n",
    "    print('annotationType: ', annotationType)\n",
    "\n",
    "    highriskGenecountsDf = highriskAnnotationsDf[highriskAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'high_risk_genes_count'})\n",
    "\n",
    "    controlGenecountsDf = controlAnnotationsDf[controlAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'control_genes_count'})\n",
    "\n",
    "    mergedGenecountsDf = controlGenecountsDf.merge(\n",
    "        highriskGenecountsDf,\n",
    "        how='left',\n",
    "        on=['Gene symbol']\n",
    "    ).fillna(0)\n",
    "\n",
    "    mergedGenecountsDf['control_genes_proportion'] = mergedGenecountsDf.control_genes_count/mergedGenecountsDf.control_genes_count.sum()\n",
    "\n",
    "    mergedGenecountsDf['expected_genes_count'] = mergedGenecountsDf.control_genes_proportion * mergedGenecountsDf.high_risk_genes_count.sum()\n",
    "\n",
    "    filteredGenecountsDf = mergedGenecountsDf[(mergedGenecountsDf.high_risk_genes_count >= 5) & (mergedGenecountsDf.expected_genes_count >= 5) & (mergedGenecountsDf.high_risk_genes_count > 0)]\n",
    "\n",
    "    dfDict[annotationType] = filteredGenecountsDf\n",
    "\n",
    "    if(filteredGenecountsDf.shape[0] < 2):\n",
    "        print('Not sufficient data for the test')\n",
    "        continue\n",
    "\n",
    "    chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredGenecountsDf.high_risk_genes_count, filteredGenecountsDf.expected_genes_count), correction=True)\n",
    "    significant = p < 0.05  # 5% significance level\n",
    "    print(chi2, p, significant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annotationType:  VIRULENCE\n",
      "10.869132557035037 0.9495459165974602 False\n",
      "annotationType:  AMR\n",
      "1.399489055858899 0.9658894901013364 False\n",
      "annotationType:  STRESS\n",
      "46.65687621578695 1.7976381917800828e-09 True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1024302/277457695.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredGenecountsDf['control_genes_proportion'] = filteredGenecountsDf.control_genes_count/filteredGenecountsDf.control_genes_count.sum()\n",
      "/tmp/ipykernel_1024302/277457695.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredGenecountsDf['expected_genes_count'] = filteredGenecountsDf.control_genes_proportion * filteredGenecountsDf.high_risk_genes_count.sum()\n",
      "/tmp/ipykernel_1024302/277457695.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredGenecountsDf['control_genes_proportion'] = filteredGenecountsDf.control_genes_count/filteredGenecountsDf.control_genes_count.sum()\n",
      "/tmp/ipykernel_1024302/277457695.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredGenecountsDf['expected_genes_count'] = filteredGenecountsDf.control_genes_proportion * filteredGenecountsDf.high_risk_genes_count.sum()\n",
      "/tmp/ipykernel_1024302/277457695.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredGenecountsDf['control_genes_proportion'] = filteredGenecountsDf.control_genes_count/filteredGenecountsDf.control_genes_count.sum()\n",
      "/tmp/ipykernel_1024302/277457695.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredGenecountsDf['expected_genes_count'] = filteredGenecountsDf.control_genes_proportion * filteredGenecountsDf.high_risk_genes_count.sum()\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "\n",
    "dfDict = {}\n",
    "\n",
    "for annotationType in ['VIRULENCE', 'AMR', 'STRESS']:\n",
    "# for annotationType in ['AMR']:\n",
    "\n",
    "    print('annotationType: ', annotationType)\n",
    "\n",
    "    highriskGenecountsDf = highriskAnnotationsDf[highriskAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'high_risk_genes_count'})\n",
    "\n",
    "    controlGenecountsDf = controlAnnotationsDf[controlAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'control_genes_count'})\n",
    "\n",
    "    mergedGenecountsDf = controlGenecountsDf.merge(\n",
    "        highriskGenecountsDf,\n",
    "        how='left',\n",
    "        on=['Gene symbol']\n",
    "    ).fillna(0)\n",
    "\n",
    "    filteredGenecountsDf = mergedGenecountsDf[(mergedGenecountsDf.high_risk_genes_count >= 5) & (mergedGenecountsDf.high_risk_genes_count > 0)]\n",
    "\n",
    "    filteredGenecountsDf['control_genes_proportion'] = filteredGenecountsDf.control_genes_count/filteredGenecountsDf.control_genes_count.sum()\n",
    "\n",
    "    filteredGenecountsDf['expected_genes_count'] = filteredGenecountsDf.control_genes_proportion * filteredGenecountsDf.high_risk_genes_count.sum()\n",
    "\n",
    "    dfDict[annotationType] = filteredGenecountsDf\n",
    "\n",
    "    if(filteredGenecountsDf.shape[0] < 2):\n",
    "        print('Not sufficient data for the test')\n",
    "        continue\n",
    "\n",
    "    chi2, p = stats.chisquare(f_obs=filteredGenecountsDf.high_risk_genes_count, f_exp=filteredGenecountsDf.expected_genes_count)\n",
    "    significant = p < 0.05  # 5% significance level\n",
    "    print(chi2, p, significant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>aur</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.073038</td>\n",
       "      <td>25.490102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>hld</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.073038</td>\n",
       "      <td>25.490102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>hlgA</td>\n",
       "      <td>531</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.072491</td>\n",
       "      <td>25.299522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>hlgB</td>\n",
       "      <td>531</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.072491</td>\n",
       "      <td>25.299522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>hlgC</td>\n",
       "      <td>530</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.072355</td>\n",
       "      <td>25.251877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>icaC</td>\n",
       "      <td>536</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.073174</td>\n",
       "      <td>25.537747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>lukD</td>\n",
       "      <td>244</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.033311</td>\n",
       "      <td>11.625392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>lukE</td>\n",
       "      <td>353</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.048191</td>\n",
       "      <td>16.818703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sak</td>\n",
       "      <td>417</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.056928</td>\n",
       "      <td>19.867986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>scn</td>\n",
       "      <td>514</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.070171</td>\n",
       "      <td>24.489556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>sei</td>\n",
       "      <td>251</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.034266</td>\n",
       "      <td>11.958908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>sel</td>\n",
       "      <td>53</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.007235</td>\n",
       "      <td>2.525188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>sel26</td>\n",
       "      <td>216</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.029488</td>\n",
       "      <td>10.291331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>selX</td>\n",
       "      <td>248</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.033857</td>\n",
       "      <td>11.815973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>sem</td>\n",
       "      <td>252</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.034403</td>\n",
       "      <td>12.006553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>sen</td>\n",
       "      <td>250</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.034130</td>\n",
       "      <td>11.911263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>seo</td>\n",
       "      <td>208</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.028396</td>\n",
       "      <td>9.910171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>seu</td>\n",
       "      <td>250</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.034130</td>\n",
       "      <td>11.911263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>splA</td>\n",
       "      <td>314</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.042867</td>\n",
       "      <td>14.960546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>splB</td>\n",
       "      <td>309</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.042184</td>\n",
       "      <td>14.722321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>splE</td>\n",
       "      <td>248</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.033857</td>\n",
       "      <td>11.815973</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "16          aur                  535                   26.0   \n",
       "44          hld                  535                   26.0   \n",
       "45         hlgA                  531                   26.0   \n",
       "46         hlgB                  531                   26.0   \n",
       "47         hlgC                  530                   26.0   \n",
       "51         icaC                  536                   27.0   \n",
       "68         lukD                  244                    8.0   \n",
       "69         lukE                  353                   13.0   \n",
       "91          sak                  417                   23.0   \n",
       "93          scn                  514                   24.0   \n",
       "102         sei                  251                   13.0   \n",
       "105         sel                   53                    5.0   \n",
       "106       sel26                  216                   14.0   \n",
       "112        selX                  248                   13.0   \n",
       "114         sem                  252                   13.0   \n",
       "115         sen                  250                   13.0   \n",
       "117         seo                  208                   11.0   \n",
       "123         seu                  250                   13.0   \n",
       "128        splA                  314                   10.0   \n",
       "129        splB                  309                   10.0   \n",
       "130        splE                  248                    9.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  \n",
       "16                   0.073038             25.490102  \n",
       "44                   0.073038             25.490102  \n",
       "45                   0.072491             25.299522  \n",
       "46                   0.072491             25.299522  \n",
       "47                   0.072355             25.251877  \n",
       "51                   0.073174             25.537747  \n",
       "68                   0.033311             11.625392  \n",
       "69                   0.048191             16.818703  \n",
       "91                   0.056928             19.867986  \n",
       "93                   0.070171             24.489556  \n",
       "102                  0.034266             11.958908  \n",
       "105                  0.007235              2.525188  \n",
       "106                  0.029488             10.291331  \n",
       "112                  0.033857             11.815973  \n",
       "114                  0.034403             12.006553  \n",
       "115                  0.034130             11.911263  \n",
       "117                  0.028396              9.910171  \n",
       "123                  0.034130             11.911263  \n",
       "128                  0.042867             14.960546  \n",
       "129                  0.042184             14.722321  \n",
       "130                  0.033857             11.815973  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfDict['VIRULENCE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>blaI</td>\n",
       "      <td>451</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.173863</td>\n",
       "      <td>21.211257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>blaR1</td>\n",
       "      <td>313</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.120663</td>\n",
       "      <td>14.720894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>blaZ</td>\n",
       "      <td>379</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.146106</td>\n",
       "      <td>17.824981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>fosB</td>\n",
       "      <td>296</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.114109</td>\n",
       "      <td>13.921357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>mecA</td>\n",
       "      <td>85</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.032768</td>\n",
       "      <td>3.997687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>mepA</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.206245</td>\n",
       "      <td>25.161912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>tet(38)</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.206245</td>\n",
       "      <td>25.161912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "109        blaI                  451                   21.0   \n",
       "244       blaR1                  313                   14.0   \n",
       "288        blaZ                  379                   17.0   \n",
       "339        fosB                  296                   12.0   \n",
       "357        mecA                   85                    6.0   \n",
       "363        mepA                  535                   26.0   \n",
       "420     tet(38)                  535                   26.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  \n",
       "109                  0.173863             21.211257  \n",
       "244                  0.120663             14.720894  \n",
       "288                  0.146106             17.824981  \n",
       "339                  0.114109             13.921357  \n",
       "357                  0.032768              3.997687  \n",
       "363                  0.206245             25.161912  \n",
       "420                  0.206245             25.161912  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfDict['AMR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>arsB</td>\n",
       "      <td>284</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.102122</td>\n",
       "      <td>5.923049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>arsC</td>\n",
       "      <td>1233</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.443366</td>\n",
       "      <td>25.715210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>arsR</td>\n",
       "      <td>438</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.157497</td>\n",
       "      <td>9.134844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cadD</td>\n",
       "      <td>335</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.120460</td>\n",
       "      <td>6.986695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>lmrS</td>\n",
       "      <td>491</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.176555</td>\n",
       "      <td>10.240201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "1         arsB                  284                    6.0   \n",
       "2         arsC                 1233                    6.0   \n",
       "6         arsR                  438                    6.0   \n",
       "12        cadD                  335                   15.0   \n",
       "25        lmrS                  491                   25.0   \n",
       "\n",
       "    control_genes_proportion  expected_genes_count  \n",
       "1                   0.102122              5.923049  \n",
       "2                   0.443366             25.715210  \n",
       "6                   0.157497              9.134844  \n",
       "12                  0.120460              6.986695  \n",
       "25                  0.176555             10.240201  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfDict['STRESS']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Old Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2     2.756334\n",
       "21    3.243666\n",
       "Name: expected_genes_count, dtype: float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filteredGenecountsDf.high_risk_genes_count.sum() / filteredGenecountsDf.expected_genes_count.sum() * filteredGenecountsDf.expected_genes_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aac(3)-IId</td>\n",
       "      <td>76</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.612665</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aac(3)-IIe</td>\n",
       "      <td>19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.153166</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(3)-IIg</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aac(3)-IVa</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>aac(6')</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002001</td>\n",
       "      <td>0.290210</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>vanX-B</td>\n",
       "      <td>135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007505</td>\n",
       "      <td>1.088286</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>vanXY-C</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.064491</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>vanY-A</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.161228</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>vanY-B</td>\n",
       "      <td>134</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007450</td>\n",
       "      <td>1.080225</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>vanZ-A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>458 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "0    aac(3)-IId                   76                    0.0   \n",
       "1    aac(3)-IIe                   19                    0.0   \n",
       "2    aac(3)-IIg                    1                    0.0   \n",
       "3    aac(3)-IVa                    1                    0.0   \n",
       "4       aac(6')                   36                    0.0   \n",
       "..          ...                  ...                    ...   \n",
       "453      vanX-B                  135                    0.0   \n",
       "454     vanXY-C                    8                    0.0   \n",
       "455      vanY-A                   20                    0.0   \n",
       "456      vanY-B                  134                    0.0   \n",
       "457      vanZ-A                    1                    0.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "0                    0.004225              0.612665                  100.0  \n",
       "1                    0.001056              0.153166                  100.0  \n",
       "2                    0.000056              0.008061                  100.0  \n",
       "3                    0.000056              0.008061                  100.0  \n",
       "4                    0.002001              0.290210                  100.0  \n",
       "..                        ...                   ...                    ...  \n",
       "453                  0.007505              1.088286                  100.0  \n",
       "454                  0.000445              0.064491                  100.0  \n",
       "455                  0.001112              0.161228                  100.0  \n",
       "456                  0.007450              1.080225                  100.0  \n",
       "457                  0.000056              0.008061                  100.0  \n",
       "\n",
       "[458 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf['percentage_difference'] = (mergedGenecountsDf.expected_genes_count - mergedGenecountsDf.high_risk_genes_count)/mergedGenecountsDf.expected_genes_count * 100\n",
    "mergedGenecountsDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aac(3)-IId</td>\n",
       "      <td>76</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.612665</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aac(3)-IIe</td>\n",
       "      <td>19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.153166</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(3)-IIg</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aac(3)-IVa</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>aac(6')</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002001</td>\n",
       "      <td>0.290210</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>vanX-B</td>\n",
       "      <td>135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007505</td>\n",
       "      <td>1.088286</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>vanXY-C</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.064491</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>vanY-A</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.161228</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>vanY-B</td>\n",
       "      <td>134</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007450</td>\n",
       "      <td>1.080225</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>vanZ-A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>442 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "0    aac(3)-IId                   76                    0.0   \n",
       "1    aac(3)-IIe                   19                    0.0   \n",
       "2    aac(3)-IIg                    1                    0.0   \n",
       "3    aac(3)-IVa                    1                    0.0   \n",
       "4       aac(6')                   36                    0.0   \n",
       "..          ...                  ...                    ...   \n",
       "453      vanX-B                  135                    0.0   \n",
       "454     vanXY-C                    8                    0.0   \n",
       "455      vanY-A                   20                    0.0   \n",
       "456      vanY-B                  134                    0.0   \n",
       "457      vanZ-A                    1                    0.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "0                    0.004225              0.612665                  100.0  \n",
       "1                    0.001056              0.153166                  100.0  \n",
       "2                    0.000056              0.008061                  100.0  \n",
       "3                    0.000056              0.008061                  100.0  \n",
       "4                    0.002001              0.290210                  100.0  \n",
       "..                        ...                   ...                    ...  \n",
       "453                  0.007505              1.088286                  100.0  \n",
       "454                  0.000445              0.064491                  100.0  \n",
       "455                  0.001112              0.161228                  100.0  \n",
       "456                  0.007450              1.080225                  100.0  \n",
       "457                  0.000056              0.008061                  100.0  \n",
       "\n",
       "[442 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf[mergedGenecountsDf.high_risk_genes_count == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>fusC</td>\n",
       "      <td>12</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.096737</td>\n",
       "      <td>-1967.471264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>tet(K)</td>\n",
       "      <td>19</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.153166</td>\n",
       "      <td>-1205.771325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>mecA</td>\n",
       "      <td>85</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.004726</td>\n",
       "      <td>0.685217</td>\n",
       "      <td>-775.634888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>blaPC1</td>\n",
       "      <td>65</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.003614</td>\n",
       "      <td>0.523990</td>\n",
       "      <td>-663.374005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>mepA</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.029744</td>\n",
       "      <td>4.312837</td>\n",
       "      <td>-502.851434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>tet(38)</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.029744</td>\n",
       "      <td>4.312837</td>\n",
       "      <td>-502.851434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>mecR1</td>\n",
       "      <td>64</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.003558</td>\n",
       "      <td>0.515928</td>\n",
       "      <td>-481.476293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>blaI</td>\n",
       "      <td>451</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.025074</td>\n",
       "      <td>3.635681</td>\n",
       "      <td>-477.608380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>blaZ</td>\n",
       "      <td>379</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.021071</td>\n",
       "      <td>3.055262</td>\n",
       "      <td>-456.417069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>blaR1</td>\n",
       "      <td>313</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.017401</td>\n",
       "      <td>2.523211</td>\n",
       "      <td>-454.848518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>fosB</td>\n",
       "      <td>296</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.016456</td>\n",
       "      <td>2.386168</td>\n",
       "      <td>-402.898416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>erm(A)</td>\n",
       "      <td>100</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.005560</td>\n",
       "      <td>0.806138</td>\n",
       "      <td>-396.193103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ant(9)-Ia</td>\n",
       "      <td>80</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.004448</td>\n",
       "      <td>0.644910</td>\n",
       "      <td>-365.181034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>erm(C)</td>\n",
       "      <td>55</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.003058</td>\n",
       "      <td>0.443376</td>\n",
       "      <td>-351.084639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>erm(T)</td>\n",
       "      <td>105</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.005838</td>\n",
       "      <td>0.846445</td>\n",
       "      <td>-18.141215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>tet(M)</td>\n",
       "      <td>492</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.027353</td>\n",
       "      <td>3.966198</td>\n",
       "      <td>49.573872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "341        fusC                   12                    2.0   \n",
       "428      tet(K)                   19                    2.0   \n",
       "357        mecA                   85                    6.0   \n",
       "207      blaPC1                   65                    4.0   \n",
       "363        mepA                  535                   26.0   \n",
       "420     tet(38)                  535                   26.0   \n",
       "359       mecR1                   64                    3.0   \n",
       "109        blaI                  451                   21.0   \n",
       "288        blaZ                  379                   17.0   \n",
       "244       blaR1                  313                   14.0   \n",
       "339        fosB                  296                   12.0   \n",
       "327      erm(A)                  100                    4.0   \n",
       "40    ant(9)-Ia                   80                    3.0   \n",
       "329      erm(C)                   55                    2.0   \n",
       "330      erm(T)                  105                    1.0   \n",
       "430      tet(M)                  492                    2.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "341                  0.000667              0.096737           -1967.471264  \n",
       "428                  0.001056              0.153166           -1205.771325  \n",
       "357                  0.004726              0.685217            -775.634888  \n",
       "207                  0.003614              0.523990            -663.374005  \n",
       "363                  0.029744              4.312837            -502.851434  \n",
       "420                  0.029744              4.312837            -502.851434  \n",
       "359                  0.003558              0.515928            -481.476293  \n",
       "109                  0.025074              3.635681            -477.608380  \n",
       "288                  0.021071              3.055262            -456.417069  \n",
       "244                  0.017401              2.523211            -454.848518  \n",
       "339                  0.016456              2.386168            -402.898416  \n",
       "327                  0.005560              0.806138            -396.193103  \n",
       "40                   0.004448              0.644910            -365.181034  \n",
       "329                  0.003058              0.443376            -351.084639  \n",
       "330                  0.005838              0.846445             -18.141215  \n",
       "430                  0.027353              3.966198              49.573872  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf[mergedGenecountsDf.high_risk_genes_count != 0].sort_values(by=['percentage_difference'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
