{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform Statistical Analysis for AMR genes\n",
    "\n",
    "This analysis is based on: https://www.kaggle.com/code/hamelg/python-for-data-25-chi-squared-tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get high risk patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a function to read data from the FHIR server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "\n",
    "\n",
    "def get(url):\n",
    "    response = requests.get(\n",
    "        url=url,\n",
    "        headers={\"Content-Type\": \"application/fhir+json\", \"authentication\": \"mjRmoNGW6klxaClkKhEkqi7HVYwx6NTH\"},\n",
    "    )\n",
    "    return response\n",
    "\n",
    "\n",
    "def readData(url):\n",
    "    nextUrl = url\n",
    "    data = []\n",
    "    while(nextUrl):\n",
    "        print('Reading URL: ', nextUrl)\n",
    "        response = get(nextUrl)\n",
    "        responseText = json.loads(response.text)\n",
    "        data.append(responseText)\n",
    "        nextUrl = None\n",
    "        if 'link' in responseText:\n",
    "            for link in responseText['link']:\n",
    "                if link['relation'] == 'next':\n",
    "                    nextUrl = link['url']\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain high risk patient ids from FHIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading URL:  http://10.172.235.4:8080/fhir/Patient?_has:RiskAssessment:subject:probability=ge0.5&_has:RiskAssessment:subject:probability=le1.0\n",
      "Reading URL:  http://10.172.235.4:8080/fhir?_getpages=0e7398ab-e927-4f59-bae7-cd4f3f07579b&_getpagesoffset=20&_count=20&_pretty=true&_bundletype=searchset\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['745962',\n",
       " '2150228',\n",
       " '2142899',\n",
       " '677694',\n",
       " '2202086',\n",
       " '2172200',\n",
       " '546011',\n",
       " '1013210',\n",
       " '2103171',\n",
       " '2013664',\n",
       " '2120861',\n",
       " '2448944',\n",
       " '2153643',\n",
       " '1581023',\n",
       " '2140940',\n",
       " '631550',\n",
       " '2198313',\n",
       " '2161817',\n",
       " '2107492',\n",
       " '2141593',\n",
       " '549608',\n",
       " '2160210',\n",
       " '2092580',\n",
       " '2156000',\n",
       " '637422',\n",
       " '2199705']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "[lowerRiskScore, higherRiskScore] = [0.5, 1.0]\n",
    "query = 'http://10.172.235.4:8080/fhir/Patient?_has:RiskAssessment:subject:probability=ge' + str(lowerRiskScore) + '&_has:RiskAssessment:subject:probability=le' + str(higherRiskScore)\n",
    "response = readData(query)\n",
    "\n",
    "highriskPatientIds = list(itertools.chain.from_iterable(list(map(lambda data: list(map(lambda entry: entry['resource']['id'][1:], data['entry'])), response))))\n",
    "highriskPatientIds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map tube codes for high risk patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tube_code</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>db_ID</th>\n",
       "      <th>pt_age</th>\n",
       "      <th>pt_gender</th>\n",
       "      <th>EPISODE_ID</th>\n",
       "      <th>hospital_admission</th>\n",
       "      <th>hospital_discharge</th>\n",
       "      <th>hospital_in_last_year</th>\n",
       "      <th>genome_species</th>\n",
       "      <th>species_reported</th>\n",
       "      <th>contig_number</th>\n",
       "      <th>length</th>\n",
       "      <th>tube_code_duplicate_or_old</th>\n",
       "      <th>date_of_collection</th>\n",
       "      <th>ID_number</th>\n",
       "      <th>location_additional_02</th>\n",
       "      <th>collected_from_original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH19J072</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>92</td>\n",
       "      <td>5131021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-0997</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH19J074</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus sp. D19</td>\n",
       "      <td>Streptococcus mitis group</td>\n",
       "      <td>84</td>\n",
       "      <td>2022991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-1174</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH19B003</td>\n",
       "      <td>526238.0</td>\n",
       "      <td>567AE</td>\n",
       "      <td>85</td>\n",
       "      <td>Male</td>\n",
       "      <td>640261.0</td>\n",
       "      <td>2019-02-04</td>\n",
       "      <td>2019-02-07</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>93</td>\n",
       "      <td>4942915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-02-03</td>\n",
       "      <td>19-034-1596</td>\n",
       "      <td>A-4WB;GMC</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH20A024</td>\n",
       "      <td>788941.0</td>\n",
       "      <td>3YBDS</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>13194824.0</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>110</td>\n",
       "      <td>5149157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>20-016-2112</td>\n",
       "      <td>S-ED;Emergency</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH20L041</td>\n",
       "      <td>1023735.0</td>\n",
       "      <td>2VJ78</td>\n",
       "      <td>42</td>\n",
       "      <td>Male</td>\n",
       "      <td>14210147.0</td>\n",
       "      <td>2020-12-18</td>\n",
       "      <td>2021-01-05</td>\n",
       "      <td>yes</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>46</td>\n",
       "      <td>2735216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-16</td>\n",
       "      <td>20-351-2856</td>\n",
       "      <td>Emergency Dept;nan</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3112</th>\n",
       "      <td>ALF22L076</td>\n",
       "      <td>639020.0</td>\n",
       "      <td>3S6HL</td>\n",
       "      <td>84</td>\n",
       "      <td>Male</td>\n",
       "      <td>17219899.0</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>2022-12-20</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus vestibularis</td>\n",
       "      <td>Streptococcus salivarius group</td>\n",
       "      <td>107</td>\n",
       "      <td>1938627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>22-349-2150</td>\n",
       "      <td>Ward 4WB</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>ALF22L089</td>\n",
       "      <td>655404.0</td>\n",
       "      <td>6EWBM</td>\n",
       "      <td>67</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteus columbae</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>152</td>\n",
       "      <td>3972219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-17</td>\n",
       "      <td>22-351-1465</td>\n",
       "      <td>Comm/Amb Clinic</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3114</th>\n",
       "      <td>ALF22L138</td>\n",
       "      <td>2572048.0</td>\n",
       "      <td>YM5PJ</td>\n",
       "      <td>61</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>Enterobacter cloacae complex</td>\n",
       "      <td>66</td>\n",
       "      <td>4876179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-27</td>\n",
       "      <td>22-361-0202</td>\n",
       "      <td>Ward 7 East</td>\n",
       "      <td>Hick White Lum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3115</th>\n",
       "      <td>ALF23A102</td>\n",
       "      <td>2054964.0</td>\n",
       "      <td>5DYAH</td>\n",
       "      <td>35</td>\n",
       "      <td>Female</td>\n",
       "      <td>17294287.0</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>2023-02-01</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus constellatus</td>\n",
       "      <td>Streptococcus anginosus group</td>\n",
       "      <td>54</td>\n",
       "      <td>1869828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-11</td>\n",
       "      <td>23-011-3305</td>\n",
       "      <td>ICU</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3116</th>\n",
       "      <td>ALF23B102</td>\n",
       "      <td>2015952.0</td>\n",
       "      <td>57E2W</td>\n",
       "      <td>82</td>\n",
       "      <td>Female</td>\n",
       "      <td>17281239.0</td>\n",
       "      <td>2023-01-08</td>\n",
       "      <td>2023-02-16</td>\n",
       "      <td>no</td>\n",
       "      <td>Acinetobacter colistiniresistens</td>\n",
       "      <td>Acinetobacter species</td>\n",
       "      <td>233</td>\n",
       "      <td>3952992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-15</td>\n",
       "      <td>23-046-3310</td>\n",
       "      <td>ICU</td>\n",
       "      <td>PICC Line White</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3117 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      tube_code  PATIENT_ID  db_ID  pt_age pt_gender  EPISODE_ID  \\\n",
       "0      AH19J072    213972.0  3JNQS      64      Male  12931474.0   \n",
       "1      AH19J074    213972.0  3JNQS      64      Male  12931474.0   \n",
       "2      AH19B003    526238.0  567AE      85      Male    640261.0   \n",
       "3      AH20A024    788941.0  3YBDS      58    Female  13194824.0   \n",
       "4      AH20L041   1023735.0  2VJ78      42      Male  14210147.0   \n",
       "...         ...         ...    ...     ...       ...         ...   \n",
       "3112  ALF22L076    639020.0  3S6HL      84      Male  17219899.0   \n",
       "3113  ALF22L089    655404.0  6EWBM      67    Female         NaN   \n",
       "3114  ALF22L138   2572048.0  YM5PJ      61      Male         NaN   \n",
       "3115  ALF23A102   2054964.0  5DYAH      35    Female  17294287.0   \n",
       "3116  ALF23B102   2015952.0  57E2W      82    Female  17281239.0   \n",
       "\n",
       "     hospital_admission hospital_discharge hospital_in_last_year  \\\n",
       "0            2019-10-17         2019-11-29                   yes   \n",
       "1            2019-10-17         2019-11-29                   yes   \n",
       "2            2019-02-04         2019-02-07                    no   \n",
       "3            2020-01-16         2020-01-19                    no   \n",
       "4            2020-12-18         2021-01-05                   yes   \n",
       "...                 ...                ...                   ...   \n",
       "3112         2022-12-15         2022-12-20                   yes   \n",
       "3113                NaN                NaN                   NaN   \n",
       "3114                NaN                NaN                   NaN   \n",
       "3115         2023-01-12         2023-02-01                   yes   \n",
       "3116         2023-01-08         2023-02-16                    no   \n",
       "\n",
       "                        genome_species                species_reported  \\\n",
       "0                     Escherichia coli                Escherichia coli   \n",
       "1                Streptococcus sp. D19       Streptococcus mitis group   \n",
       "2                     Escherichia coli                Escherichia coli   \n",
       "3                     Escherichia coli                Escherichia coli   \n",
       "4                Staphylococcus aureus           Staphylococcus aureus   \n",
       "...                                ...                             ...   \n",
       "3112        Streptococcus vestibularis  Streptococcus salivarius group   \n",
       "3113                  Proteus columbae                Proteus vulgaris   \n",
       "3114         Enterobacter roggenkampii    Enterobacter cloacae complex   \n",
       "3115        Streptococcus constellatus   Streptococcus anginosus group   \n",
       "3116  Acinetobacter colistiniresistens           Acinetobacter species   \n",
       "\n",
       "      contig_number   length tube_code_duplicate_or_old date_of_collection  \\\n",
       "0                92  5131021                        NaN         2019-10-30   \n",
       "1                84  2022991                        NaN         2019-10-30   \n",
       "2                93  4942915                        NaN         2019-02-03   \n",
       "3               110  5149157                        NaN         2020-01-16   \n",
       "4                46  2735216                        NaN         2020-12-16   \n",
       "...             ...      ...                        ...                ...   \n",
       "3112            107  1938627                        NaN         2022-12-15   \n",
       "3113            152  3972219                        NaN         2022-12-17   \n",
       "3114             66  4876179                        NaN         2022-12-27   \n",
       "3115             54  1869828                        NaN         2023-01-11   \n",
       "3116            233  3952992                        NaN         2023-02-15   \n",
       "\n",
       "        ID_number location_additional_02 collected_from_original  \n",
       "0     19-303-0997             A-7EA;HAEM                   Blood  \n",
       "1     19-303-1174             A-7EA;HAEM        Blood Peripheral  \n",
       "2     19-034-1596              A-4WB;GMC                   Blood  \n",
       "3     20-016-2112         S-ED;Emergency            Blood Venous  \n",
       "4     20-351-2856     Emergency Dept;nan                   Blood  \n",
       "...           ...                    ...                     ...  \n",
       "3112  22-349-2150               Ward 4WB        Blood Peripheral  \n",
       "3113  22-351-1465        Comm/Amb Clinic                   Blood  \n",
       "3114  22-361-0202            Ward 7 East          Hick White Lum  \n",
       "3115  23-011-3305                    ICU            Blood Venous  \n",
       "3116  23-046-3310                    ICU         PICC Line White  \n",
       "\n",
       "[3117 rows x 18 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "mappingDf = pd.read_csv(os.environ['GENOMICS_DATA_BASE'] + '/patient_tube_id_mapping_full.tsv', sep='\\t')\n",
    "mappingDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AH20I057',\n",
       " 'AH21C022',\n",
       " 'AH18J081',\n",
       " 'AH18J080',\n",
       " 'AH19L077',\n",
       " 'AH20F012',\n",
       " 'AH19J028',\n",
       " 'AH21A034',\n",
       " 'AH20H023',\n",
       " 'AH19K013',\n",
       " 'AH20L008',\n",
       " 'AH19E069',\n",
       " 'AH20L085',\n",
       " 'AH20E051',\n",
       " 'AH20C005',\n",
       " 'AH18K024',\n",
       " 'AH20K063',\n",
       " 'AH19C012',\n",
       " 'AH19B014',\n",
       " 'AH19A043',\n",
       " 'AH19B026',\n",
       " 'AH19C070',\n",
       " 'AH20I042',\n",
       " 'AH18J055',\n",
       " 'AH19F055',\n",
       " 'AH20G044',\n",
       " 'AH21A036']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highriskTubecodes = list(mappingDf[mappingDf.PATIENT_ID.isin([int(patientId) for patientId in highriskPatientIds])].tube_code)\n",
    "highriskTubecodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((597, 23), (57018, 23))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "highriskAnnotationsDfList = []\n",
    "controlAnnotationsDfList = []\n",
    "\n",
    "gffDir = os.environ['GENOMICS_DATA_BASE'] + '/amrfinder'\n",
    "\n",
    "for fileName in os.listdir(gffDir):\n",
    "\n",
    "    tubeCode = fileName.split('.')[0].split('_')[0]\n",
    "\n",
    "    amrResultsDf = pd.read_csv(\n",
    "        gffDir + '/' + fileName,\n",
    "        sep='\\t',\n",
    "    )\n",
    "    amrResultsDf['tube_code'] = tubeCode\n",
    "\n",
    "    if tubeCode in highriskTubecodes:\n",
    "        highriskAnnotationsDfList.append(amrResultsDf)\n",
    "    else:\n",
    "        controlAnnotationsDfList.append(amrResultsDf)\n",
    "\n",
    "highriskAnnotationsDf = pd.concat(highriskAnnotationsDfList, ignore_index=True)\n",
    "controlAnnotationsDf = pd.concat(controlAnnotationsDfList, ignore_index=True)\n",
    "\n",
    "highriskAnnotationsDf.shape, controlAnnotationsDf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27, 2953)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(highriskAnnotationsDf.tube_code.unique()), len(controlAnnotationsDf.tube_code.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform chi-square tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Element type</th>\n",
       "      <th>Element subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VIRULENCE</td>\n",
       "      <td>VIRULENCE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>BIOCIDE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>METAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>ACID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>STRESS</td>\n",
       "      <td>HEAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Element type Element subtype\n",
       "0      VIRULENCE       VIRULENCE\n",
       "3            AMR             AMR\n",
       "7         STRESS         BIOCIDE\n",
       "18        STRESS           METAL\n",
       "29        STRESS            ACID\n",
       "220       STRESS            HEAT"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "controlAnnotationsDf[['Element type', 'Element subtype']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annotationType:  AMR\n",
      "Power_divergenceResult(statistic=735.394986353418, pvalue=2.4268692591733216e-15)\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "\n",
    "# for annotationType in ['VIRULENCE', 'AMR', 'STRESS']:\n",
    "for annotationType in ['AMR']:\n",
    "\n",
    "    print('annotationType: ', annotationType)\n",
    "\n",
    "    highriskGenecountsDf = highriskAnnotationsDf[highriskAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'high_risk_genes_count'})\n",
    "\n",
    "    controlGenecountsDf = controlAnnotationsDf[controlAnnotationsDf['Element type'] == annotationType][['Contig id', 'Gene symbol']].groupby(\n",
    "            by=['Gene symbol']\n",
    "        ).agg(\n",
    "            'count'\n",
    "        ).reset_index().rename(columns={'Contig id': 'control_genes_count'})\n",
    "\n",
    "    mergedGenecountsDf = controlGenecountsDf.merge(\n",
    "        highriskGenecountsDf,\n",
    "        how='left',\n",
    "        on=['Gene symbol']\n",
    "    ).fillna(0)\n",
    "\n",
    "    mergedGenecountsDf['control_genes_proportion'] = mergedGenecountsDf.control_genes_count/mergedGenecountsDf.control_genes_count.sum()\n",
    "\n",
    "    mergedGenecountsDf['expected_genes_count'] = mergedGenecountsDf.control_genes_proportion * mergedGenecountsDf.high_risk_genes_count.sum()\n",
    "\n",
    "    print(stats.chisquare(f_obs= mergedGenecountsDf.high_risk_genes_count, f_exp= mergedGenecountsDf.expected_genes_count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aac(3)-IId</td>\n",
       "      <td>76</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.612665</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aac(3)-IIe</td>\n",
       "      <td>19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.153166</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(3)-IIg</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aac(3)-IVa</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>aac(6')</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002001</td>\n",
       "      <td>0.290210</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>vanX-B</td>\n",
       "      <td>135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007505</td>\n",
       "      <td>1.088286</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>vanXY-C</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.064491</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>vanY-A</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.161228</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>vanY-B</td>\n",
       "      <td>134</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007450</td>\n",
       "      <td>1.080225</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>vanZ-A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>458 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "0    aac(3)-IId                   76                    0.0   \n",
       "1    aac(3)-IIe                   19                    0.0   \n",
       "2    aac(3)-IIg                    1                    0.0   \n",
       "3    aac(3)-IVa                    1                    0.0   \n",
       "4       aac(6')                   36                    0.0   \n",
       "..          ...                  ...                    ...   \n",
       "453      vanX-B                  135                    0.0   \n",
       "454     vanXY-C                    8                    0.0   \n",
       "455      vanY-A                   20                    0.0   \n",
       "456      vanY-B                  134                    0.0   \n",
       "457      vanZ-A                    1                    0.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "0                    0.004225              0.612665                  100.0  \n",
       "1                    0.001056              0.153166                  100.0  \n",
       "2                    0.000056              0.008061                  100.0  \n",
       "3                    0.000056              0.008061                  100.0  \n",
       "4                    0.002001              0.290210                  100.0  \n",
       "..                        ...                   ...                    ...  \n",
       "453                  0.007505              1.088286                  100.0  \n",
       "454                  0.000445              0.064491                  100.0  \n",
       "455                  0.001112              0.161228                  100.0  \n",
       "456                  0.007450              1.080225                  100.0  \n",
       "457                  0.000056              0.008061                  100.0  \n",
       "\n",
       "[458 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf['percentage_difference'] = (mergedGenecountsDf.expected_genes_count - mergedGenecountsDf.high_risk_genes_count)/mergedGenecountsDf.expected_genes_count * 100\n",
    "mergedGenecountsDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aac(3)-IId</td>\n",
       "      <td>76</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.612665</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aac(3)-IIe</td>\n",
       "      <td>19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.153166</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aac(3)-IIg</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aac(3)-IVa</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>aac(6')</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002001</td>\n",
       "      <td>0.290210</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>vanX-B</td>\n",
       "      <td>135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007505</td>\n",
       "      <td>1.088286</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>vanXY-C</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.064491</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>vanY-A</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.161228</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>vanY-B</td>\n",
       "      <td>134</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007450</td>\n",
       "      <td>1.080225</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>vanZ-A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008061</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>442 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "0    aac(3)-IId                   76                    0.0   \n",
       "1    aac(3)-IIe                   19                    0.0   \n",
       "2    aac(3)-IIg                    1                    0.0   \n",
       "3    aac(3)-IVa                    1                    0.0   \n",
       "4       aac(6')                   36                    0.0   \n",
       "..          ...                  ...                    ...   \n",
       "453      vanX-B                  135                    0.0   \n",
       "454     vanXY-C                    8                    0.0   \n",
       "455      vanY-A                   20                    0.0   \n",
       "456      vanY-B                  134                    0.0   \n",
       "457      vanZ-A                    1                    0.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "0                    0.004225              0.612665                  100.0  \n",
       "1                    0.001056              0.153166                  100.0  \n",
       "2                    0.000056              0.008061                  100.0  \n",
       "3                    0.000056              0.008061                  100.0  \n",
       "4                    0.002001              0.290210                  100.0  \n",
       "..                        ...                   ...                    ...  \n",
       "453                  0.007505              1.088286                  100.0  \n",
       "454                  0.000445              0.064491                  100.0  \n",
       "455                  0.001112              0.161228                  100.0  \n",
       "456                  0.007450              1.080225                  100.0  \n",
       "457                  0.000056              0.008061                  100.0  \n",
       "\n",
       "[442 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf[mergedGenecountsDf.high_risk_genes_count == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>control_genes_count</th>\n",
       "      <th>high_risk_genes_count</th>\n",
       "      <th>control_genes_proportion</th>\n",
       "      <th>expected_genes_count</th>\n",
       "      <th>percentage_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>fusC</td>\n",
       "      <td>12</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000667</td>\n",
       "      <td>0.096737</td>\n",
       "      <td>-1967.471264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>tet(K)</td>\n",
       "      <td>19</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.153166</td>\n",
       "      <td>-1205.771325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>mecA</td>\n",
       "      <td>85</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.004726</td>\n",
       "      <td>0.685217</td>\n",
       "      <td>-775.634888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>blaPC1</td>\n",
       "      <td>65</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.003614</td>\n",
       "      <td>0.523990</td>\n",
       "      <td>-663.374005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>mepA</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.029744</td>\n",
       "      <td>4.312837</td>\n",
       "      <td>-502.851434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>tet(38)</td>\n",
       "      <td>535</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.029744</td>\n",
       "      <td>4.312837</td>\n",
       "      <td>-502.851434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>mecR1</td>\n",
       "      <td>64</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.003558</td>\n",
       "      <td>0.515928</td>\n",
       "      <td>-481.476293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>blaI</td>\n",
       "      <td>451</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.025074</td>\n",
       "      <td>3.635681</td>\n",
       "      <td>-477.608380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>blaZ</td>\n",
       "      <td>379</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.021071</td>\n",
       "      <td>3.055262</td>\n",
       "      <td>-456.417069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>blaR1</td>\n",
       "      <td>313</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.017401</td>\n",
       "      <td>2.523211</td>\n",
       "      <td>-454.848518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>fosB</td>\n",
       "      <td>296</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.016456</td>\n",
       "      <td>2.386168</td>\n",
       "      <td>-402.898416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>erm(A)</td>\n",
       "      <td>100</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.005560</td>\n",
       "      <td>0.806138</td>\n",
       "      <td>-396.193103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ant(9)-Ia</td>\n",
       "      <td>80</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.004448</td>\n",
       "      <td>0.644910</td>\n",
       "      <td>-365.181034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>erm(C)</td>\n",
       "      <td>55</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.003058</td>\n",
       "      <td>0.443376</td>\n",
       "      <td>-351.084639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>erm(T)</td>\n",
       "      <td>105</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.005838</td>\n",
       "      <td>0.846445</td>\n",
       "      <td>-18.141215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>tet(M)</td>\n",
       "      <td>492</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.027353</td>\n",
       "      <td>3.966198</td>\n",
       "      <td>49.573872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene symbol  control_genes_count  high_risk_genes_count  \\\n",
       "341        fusC                   12                    2.0   \n",
       "428      tet(K)                   19                    2.0   \n",
       "357        mecA                   85                    6.0   \n",
       "207      blaPC1                   65                    4.0   \n",
       "363        mepA                  535                   26.0   \n",
       "420     tet(38)                  535                   26.0   \n",
       "359       mecR1                   64                    3.0   \n",
       "109        blaI                  451                   21.0   \n",
       "288        blaZ                  379                   17.0   \n",
       "244       blaR1                  313                   14.0   \n",
       "339        fosB                  296                   12.0   \n",
       "327      erm(A)                  100                    4.0   \n",
       "40    ant(9)-Ia                   80                    3.0   \n",
       "329      erm(C)                   55                    2.0   \n",
       "330      erm(T)                  105                    1.0   \n",
       "430      tet(M)                  492                    2.0   \n",
       "\n",
       "     control_genes_proportion  expected_genes_count  percentage_difference  \n",
       "341                  0.000667              0.096737           -1967.471264  \n",
       "428                  0.001056              0.153166           -1205.771325  \n",
       "357                  0.004726              0.685217            -775.634888  \n",
       "207                  0.003614              0.523990            -663.374005  \n",
       "363                  0.029744              4.312837            -502.851434  \n",
       "420                  0.029744              4.312837            -502.851434  \n",
       "359                  0.003558              0.515928            -481.476293  \n",
       "109                  0.025074              3.635681            -477.608380  \n",
       "288                  0.021071              3.055262            -456.417069  \n",
       "244                  0.017401              2.523211            -454.848518  \n",
       "339                  0.016456              2.386168            -402.898416  \n",
       "327                  0.005560              0.806138            -396.193103  \n",
       "40                   0.004448              0.644910            -365.181034  \n",
       "329                  0.003058              0.443376            -351.084639  \n",
       "330                  0.005838              0.846445             -18.141215  \n",
       "430                  0.027353              3.966198              49.573872  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mergedGenecountsDf[mergedGenecountsDf.high_risk_genes_count != 0].sort_values(by=['percentage_difference'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
